<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Introduction &mdash; Pykaryote 1.1.0d documentation</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.1.0d',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="Pykaryote 1.1.0d documentation" href="../index.html" />
    <link rel="next" title="Installing Pykaryote" href="installing.html" />
    <link rel="prev" title="Pykaryote Documentation" href="../index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="installing.html" title="Installing Pykaryote"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="../index.html" title="Pykaryote Documentation"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">Pykaryote 1.1.0d documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="introduction">
<span id="id1"></span><h1>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h1>
<div class="section" id="what-is-pykaryote">
<h2>What is Pykaryote<a class="headerlink" href="#what-is-pykaryote" title="Permalink to this headline">¶</a></h2>
<p>Pykaryote is a computer model for simulating the evolution of biological complexity. Pykaryote is a portmanteau of &#8216;python&#8217;, the language in which the model is implemented, and &#8216;prokaryote&#8217;, a group of organisms who lack a membrane-bound nucleus.</p>
<p>Pykaryote python package includes a suite of command line utilities for running simulations and graphing the results. Pykaryote&#8217;s main front end, petri, allows one to distribute batches of simulations over many computers using MPI.</p>
<p>Almost every aspect of a simulation can be controlled by adjusting options in Pykaryote&#8217;s extensive configuration files. By analyzing the effects of different settings, it is hoped that Pykaryote will help answer questions about what circumstances are necessary for the evolution of complexity.</p>
<div class="section" id="the-model">
<h3>The Model<a class="headerlink" href="#the-model" title="Permalink to this headline">¶</a></h3>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The following is a brief explanation of the Pykaryote model. For a more detailed description, see <a class="reference internal" href="../ref/genome.html#model-in-depth"><em>The Model in Depth</em></a>. If you just want to install and run Pykaryote, skip ahead to <a class="reference internal" href="installing.html#installing"><em>Installing Pykaryote</em></a>.</p>
</div>
<p>A pykaryote simulations consists of a group of organisms whose goal is to gather chemicals. Organisms can combine gathered chemicals into proteins which allow them to gather chemicals more efficiently. Proteins can be combined into new, more powerful proteins which are also called complexes.</p>
<p>The actions of an organism are dictated by its genome. Like biological genomes, an organism&#8217;s genome is a string of codons. An organism will read a codon, do the action associated with that codon, and then move on to the next codon. The genome is circular, so when the end is reached it loops back to the beginning.</p>
<p>Depending on what their genomes tell them to do, organisms can move about, gather chemicals and combine those chemicals into proteins&#8211;molecules that have
the potential to improve their gathering effectiveness (or they may be
useless). These proteins in turn might combine into complexes, groups of
proteins that interact, giving a greater boost to the organism (or again,
maybe nothing at all).</p>
<p>After a period of time, all the organisms are given scores based on how many
chemicals they&#8217;ve accumulated. Organisms with a high fitness are likely to have offspring. The old organisms die off and the offspring, the new generation, continue in their place. In this way we have a &#8220;reproduction of the fittest&#8221; rule.</p>
<p>Given time and the right conditions, the populations of organisms will improve, building better proteins and discovering new complexes, and as a result gathering many more chemicals than earlier generations.</p>
</div>
</div>
<div class="section" id="scientific-background">
<h2>Scientific Background<a class="headerlink" href="#scientific-background" title="Permalink to this headline">¶</a></h2>
<p>The following section provides some basic background on scientific concepts related to Pykaryote.</p>
<div class="section" id="interlocking-complexity">
<h3>Interlocking Complexity<a class="headerlink" href="#interlocking-complexity" title="Permalink to this headline">¶</a></h3>
<p>The goal of pykaryote is to study what conditions are necessary for the evolution of complexity. A working definition of interlocking complexity is &#8216;a system of several
interacting parts performing a function, where multiple parts must be
present and working properly in order for the system to perform its
function; if certain single parts are removed, the entire system is
greatly impaired or fails to function at all.&#8217;</p>
<p>Many mechanical devices display interlocking complexity. For example,
removal of a single part from a mechanical clock or music box could
prevent the entire device from functioning. Such devices typically are
designed ahead of time in great detail, and then assembled “by hand.”
There are other systems which display interlocking complexity where the
interlocking complexity seems to have self-organized or evolved. Two such
examples are modern industrial economies and biological organisms.</p>
<p>Modern industrial economies are composed of thousands of different
industries and occupations, ranging across agriculture, health care,
education, manufacturing, transportation, energy, and many others.
Economies display a great degree of interlocking complexity. If one
sub-industry (e.g. oil refining) were to stop working altogether, the
entire industry of gasoline distribution and sales would halt, and the
economy as a whole would suffer greatly until a substitute industry were
available. Many industries (e.g. rubber tire manufacturing) depend on the
oil refining industry; and in turn, oil refining depends on many other
industries (including the rubber tire manufacturing industry) in order to
work properly. Industrial economies did not achieve their complexity all
at once. Their complexity built up slowly over time from much simpler
economies.</p>
<p>Biological organisms display interlocking complexity both at the organismal
level and at the biochemical level. One famous example of the later is the
Kreb’s cycle, a sequence of catalytic chemical reactions important to all
organisms that use oxygen. Numerous enzymes are vital to the proper
functioning of this cycle; if one of them were removed from the cell, the
cycle would cease working. One of the outstanding questions of
evolutionary biology is how such systems of interlocking complexity could
have evolved through the Darwinian mechanisms of mutation and natural
selection. One of the central arguments of the Intelligent Design movement
is to dispute whether the evolution of interlocking complexity is possible
at all.</p>
</div>
<div class="section" id="evolution">
<h3>Evolution<a class="headerlink" href="#evolution" title="Permalink to this headline">¶</a></h3>
<p>Here is a vastly oversimplified model of biological evolution: each gene
makes a single protein, each protein has a single function, and the only
mutations are point mutations. Under such a model of evolution, it is
vastly improbable that interlocking complexity like the Kreb’s cycle could
evolve. But real biological evolution is more interesting. Many proteins
have multiple functions (<strong>multitasking</strong>). Some important functions in
cells are performed, or could be performed, in several ways
(<strong>redundancy</strong>). There are many kinds of mutation; to name just some
examples of mutation: gene duplication, horizontal gene transfer,
shuffling of genes, changes in gene regulation, and alternative splicing.
<strong>Neutral drift</strong>, producing <strong>genetic variability</strong> in populations, and
<strong>changing environments</strong> also play important roles in the evolution of
complexity. This combination can produce <strong>exaption</strong>, that is, a protein
or a trait serves one function, but is recruited to take on an additional
function. For example, when an environment changes, a phenotypic feature
which has some variability within a population, but is not strongly
selected prior to the environmental change, can suddenly become strongly
selected. Because of the change in environment, a pre-existing feature of
an organism takes on a <strong>novel function</strong> without losing its original
function. Such situations could make gene duplication and further
mutations an important way to improve fitness. Combining these features of
biological evolution suggests some ways to model the evolution of
interlocking complexity.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Introduction</a><ul>
<li><a class="reference internal" href="#what-is-pykaryote">What is Pykaryote</a><ul>
<li><a class="reference internal" href="#the-model">The Model</a></li>
</ul>
</li>
<li><a class="reference internal" href="#scientific-background">Scientific Background</a><ul>
<li><a class="reference internal" href="#interlocking-complexity">Interlocking Complexity</a></li>
<li><a class="reference internal" href="#evolution">Evolution</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="../index.html"
                        title="previous chapter">Pykaryote Documentation</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="installing.html"
                        title="next chapter">Installing Pykaryote</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/tut/intro.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="installing.html" title="Installing Pykaryote"
             >next</a> |</li>
        <li class="right" >
          <a href="../index.html" title="Pykaryote Documentation"
             >previous</a> |</li>
        <li><a href="../index.html">Pykaryote 1.1.0d documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2011 - 2013, James Lamine and Ethan Van Andel.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b1.
    </div>
  </body>
</html>