<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Simulation Configuration Files &mdash; Pykaryote 1.1.0d documentation</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.1.0d',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="Pykaryote 1.1.0d documentation" href="../index.html" />
    <link rel="next" title="Analyzer Data File Format" href="data-format.html" />
    <link rel="prev" title="The Model in Depth" href="genome.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="data-format.html" title="Analyzer Data File Format"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="genome.html" title="The Model in Depth"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">Pykaryote 1.1.0d documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="simulation-configuration-files">
<span id="config-files"></span><span id="config"></span><h1>Simulation Configuration Files<a class="headerlink" href="#simulation-configuration-files" title="Permalink to this headline">¶</a></h1>
<p>This section describes the simulation configuration file options.</p>
<p>Almost every aspect of a simulation can be modified by options in the simulation configuration file.</p>
<p>The default configuration values are in <tt class="docutils literal"><span class="pre">master.cfg</span></tt>, which should not be modified unless new configuration options are added to Pykaryote. <tt class="docutils literal"><span class="pre">sim.cfg</span></tt> is the default configuration file used with the command line utilities.</p>
<p>Configuration files are read by <a class="reference internal" href="utils/globals.html#module-pykaryote.utils.globals" title="pykaryote.utils.globals"><tt class="xref py py-mod docutils literal"><span class="pre">globals</span></tt></a>.</p>
<p>Configuration files use the INI file format, although the <tt class="docutils literal"><span class="pre">.cfg</span></tt> extension is used by convention.</p>
<div class="section" id="configuration-options">
<span id="vars"></span><h2>Configuration Options<a class="headerlink" href="#configuration-options" title="Permalink to this headline">¶</a></h2>
<p>The following is a list of the sections and options available in a simulation configuration file.</p>
<p>Default values indicate those present in <tt class="docutils literal"><span class="pre">master.cfg</span></tt>.</p>
<div class="section" id="global">
<h3>global<a class="headerlink" href="#global" title="Permalink to this headline">¶</a></h3>
<dl class="data">
<dt id="chem_list">
<tt class="descname">chem_list</tt><a class="headerlink" href="#chem_list" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = [5,0]</em></p>
<p>A list of the chemicals becoming available at each level of complexity.
E.g. <tt class="docutils literal"><span class="pre">[4,2,0,1]</span></tt> means there are 4 chemicals that can be gathered with no
complexes, 2 that require at least a protein, and 1 that requires
complexes of length &gt;= 3.</p>
</dd></dl>

<dl class="data">
<dt id="num_organisms">
<tt class="descname">num_organisms</tt><a class="headerlink" href="#num_organisms" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 50</em></p>
<p>The number of organisms in the population.</p>
</dd></dl>

<dl class="data">
<dt id="generation_limit">
<tt class="descname">generation_limit</tt><a class="headerlink" href="#generation_limit" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 1000</em></p>
<p>The number of generations the simulation will run for. A value of 0 means there is no limit.</p>
</dd></dl>

<dl class="data">
<dt id="runtime_limit">
<tt class="descname">runtime_limit</tt><a class="headerlink" href="#runtime_limit" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 480.0</em></p>
<p>The maximum runtime for a simulation, in seconds. If a simulation is estimated to take longer than this, it exits prematurely. A value of 0.0 means there is not limit.</p>
</dd></dl>

<dl class="data">
<dt id="complexity_limit">
<tt class="descname">complexity_limit</tt><a class="headerlink" href="#complexity_limit" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 0.0</em></p>
<p>Simulations are stopped once the average complexity reaches this value. A value of 0.0 means there is no limit.</p>
</dd></dl>

<dl class="data">
<dt id="generation_length">
<tt class="descname">generation_length</tt><a class="headerlink" href="#generation_length" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 10000</em></p>
<p>The number of codons each organism reads per generation.</p>
</dd></dl>

<dl class="data">
<dt id="org_step">
<tt class="descname">org_step</tt><a class="headerlink" href="#org_step" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 1000</em></p>
<p>The number of codons read in one chunk. This is an
optimization variable. It has no effect on the outcome of the simulation,
but for values lower than 1000 performance suffers.</p>
<p>Note that this means that organism 1 gets 1000 actions before organism 2
and so forth. Any modification (such as limited resources or pollution)
that allows organisms to interfere with eachother within a generation
must take this into account. It also limits the step size of features
such as diffusion. It should be a factor of <a class="reference internal" href="#generation_length" title="generation_length"><tt class="xref py py-data docutils literal"><span class="pre">generation_length</span></tt></a>.</p>
</dd></dl>

<dl class="data">
<dt id="log_step">
<tt class="descname">log_step</tt><a class="headerlink" href="#log_step" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 20</em></p>
<p>The number of generations between logs.</p>
</dd></dl>

<dl class="data">
<dt id="save_step">
<tt class="descname">save_step</tt><a class="headerlink" href="#save_step" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 100</em></p>
<p>The number of generations between state saves.</p>
</dd></dl>

<dl class="data">
<dt id="termination_size">
<tt class="descname">termination_size</tt><a class="headerlink" href="#termination_size" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 1000000</em></p>
<p>Runs that reach large complexity often invent massive numbers of
complexes. These complexes must be stored, and it&#8217;s fairly easy for
them to overflow the main memory capacity and essentially stop the
system. This variable sets a limit on the number of complexes invented
by a simulation. If a simulation exceeds this number, it will save
itself and stop running. It can always be recovered later by loading
(see <a class="reference internal" href="sim/sim.html#loading"><em>Loading Simulations</em></a>), and wont terminate again.</p>
</dd></dl>

<dl class="data">
<dt id="selection_exponent">
<tt class="descname">selection_exponent</tt><a class="headerlink" href="#selection_exponent" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 3.0</em></p>
<p>All organism fitnesses are raised to this power to compute their
reproduction probability. Higher values correspond to harsher
selection. There are <a class="reference internal" href="#num_organisms" title="num_organisms"><tt class="xref py py-data docutils literal"><span class="pre">num_organisms</span></tt></a> reproduction opportunities
per generation. If selection exponent = <span class="math">\(E\)</span> For each
opportunity, an organism <span class="math">\(org\)</span> has probability of reproducing
<span class="math">\(P(reproduce) = \frac{org.fit^E}{\sum_{org_i \in Orgs}
org_i.fit^E}\)</span></p>
</dd></dl>

<dl class="data">
<dt id="randomseed">
<tt class="descname">randomseed</tt><a class="headerlink" href="#randomseed" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 0</em></p>
<p>If <tt class="docutils literal"><span class="pre">0</span></tt>, <tt class="docutils literal"><span class="pre">random.random</span></tt> and <tt class="docutils literal"><span class="pre">numpy.random</span></tt> are each seeded randomly.
Otherwise, this value must be a positive integer, and it is used as the
seed, to provide deterministic behavior.  Either way, the seed that is used
is saved to <tt class="docutils literal"><span class="pre">&lt;data_folder&gt;/seed</span></tt>.</p>
</dd></dl>

<dl class="data">
<dt id="mean_to_aggregate">
<tt class="descname">mean_to_aggregate</tt><a class="headerlink" href="#mean_to_aggregate" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 0.8</em></p>
<p>When aggregating multiple runs of a simulation, outliers are ignored by
taking the middle <tt class="docutils literal"><span class="pre">mean_to_aggregate</span></tt> percent of a batch. With the default of 0.8, the lower and upper 10 percent of data are ignored.</p>
</dd></dl>

</div>
<div class="section" id="environment">
<h3>Environment<a class="headerlink" href="#environment" title="Permalink to this headline">¶</a></h3>
<dl class="data">
<dt id="rows">
<tt class="descname">rows</tt><a class="headerlink" href="#rows" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 1</em></p>
<p>The number of rows of cells in the Environment</p>
</dd></dl>

<dl class="data">
<dt id="columns">
<tt class="descname">columns</tt><a class="headerlink" href="#columns" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 1</em></p>
<p>The number of columns of cells in the Environment</p>
</dd></dl>

<dl class="data">
<dt id="wrap">
<tt class="descname">wrap</tt><a class="headerlink" href="#wrap" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = True</em></p>
<p>If True, coordinates in the environment wrap and the environment behaves
like a torus. For example, in a 10 x 10 environment, [0,5] is adjacent to
[9,5].</p>
</dd></dl>

<dl class="data">
<dt id="diffusion">
<tt class="descname">diffusion</tt><a class="headerlink" href="#diffusion" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = False</em></p>
<p>If True, resources in the environment diffuse between cells.</p>
<p>NOT FULLY IMPLEMENTED</p>
</dd></dl>

<dl class="data">
<dt id="diffusion_step">
<tt class="descname">diffusion_step</tt><a class="headerlink" href="#diffusion_step" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 1000</em></p>
<p>The number of organism instructions between diffusion operations.
This is mainly necessary for efficiency (and should be a multiple of <a class="reference internal" href="#org_step" title="org_step"><tt class="xref py py-data docutils literal"><span class="pre">org_step</span></tt></a>)</p>
</dd></dl>

<dl class="data">
<dt id="diffusion_coefficient">
<tt class="descname">diffusion_coefficient</tt><a class="headerlink" href="#diffusion_coefficient" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = .01</em></p>
<p>The amount of each chemical that diffuse in each direction.</p>
</dd></dl>

<dl class="data">
<dt id="grid_type">
<tt class="descname">grid_type</tt><a class="headerlink" href="#grid_type" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = flat</em></p>
<p>The type of resource grid for the environment. Options are:</p>
<ul class="simple">
<li>flat - equal concentrations of resources in all cells</li>
<li>gaussian - chemicals appear in clusters as specified by <tt class="docutils literal"><span class="pre">gaussians</span></tt>. See <a class="reference internal" href="#gaussians" title="gaussians"><tt class="xref py py-data docutils literal"><span class="pre">gaussians</span></tt></a></li>
</ul>
</dd></dl>

<dl class="data">
<dt id="base_concentration">
<tt class="descname">base_concentration</tt><a class="headerlink" href="#base_concentration" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 1.0</em></p>
<p>The base concentration of chemicals in each square. Applies to all grid
types.</p>
</dd></dl>

<dl class="data">
<dt id="gaussians">
<tt class="descname">gaussians</tt><a class="headerlink" href="#gaussians" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = [[0, (10, 10), 15], [1, (40, 40), 15], [2, (10, 40), 15], [3, (40, 10), 15], [4, (25, 25), 15]]</em></p>
<p>Specifies radial mounds of chemicals to place in the environment. This setting is only used if <tt class="docutils literal"><span class="pre">grid_type</span></tt> is set to <tt class="docutils literal"><span class="pre">gaussian</span></tt> (see <a class="reference internal" href="#grid_type" title="grid_type"><tt class="xref py py-data docutils literal"><span class="pre">grid_type</span></tt></a>).</p>
<p>The setting is a list of gaussians, where each gaussian is specified as a
list containing three items:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="p">[</span><span class="n">chemical</span><span class="p">,</span> <span class="p">(</span><span class="n">x_center</span><span class="p">,</span> <span class="n">y_center</span><span class="p">),</span> <span class="n">radius</span><span class="p">]</span>
</pre></div>
</div>
<p>Even though a radius is specified, radial mounds are continuous. The radius denotes the point as which the amount of chemical becomes trivial.</p>
</dd></dl>

</div>
<div class="section" id="organism">
<h3>Organism<a class="headerlink" href="#organism" title="Permalink to this headline">¶</a></h3>
<dl class="data">
<dt id="gather_proportion">
<tt class="descname">gather_proportion</tt><a class="headerlink" href="#gather_proportion" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 1.0</em></p>
<p>The proportion of the available chemical gathered with each operation (by
an organism with function[chemical] = 1)</p>
</dd></dl>

<dl class="data">
<dt id="chemical_fitnesses">
<tt class="descname">chemical_fitnesses</tt><a class="headerlink" href="#chemical_fitnesses" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = [(&#8220;root&#8221;,4),(&#8220;root&#8221;,4),(&#8220;root&#8221;,4),(&#8220;root&#8221;,4),(&#8220;root&#8221;,4)]</em></p>
<p>A vector with the fitness function for each chemical. Each element must
be a tuple of one of the following forms:</p>
<ul class="simple">
<li>(&#8220;limitlog&#8221;, <span class="math">\(c\)</span>, <span class="math">\(p\)</span>) - <span class="math">\(x\)</span> of a given
chemical is worth <span class="math">\((x^{p}-1)*c/p\)</span></li>
<li>(&#8220;log&#8221;, <span class="math">\(c\)</span>) - <span class="math">\(x\)</span> of a given power is worth <span class="math">\(\log_{2}(x+1)\)</span></li>
<li>(&#8220;linear&#8221;, <span class="math">\(c\)</span>) - <span class="math">\(x\)</span> of a chemical is simply worth <span class="math">\(x\)</span></li>
<li>(&#8220;root&#8221;, <span class="math">\(r\)</span>) - <span class="math">\(x\)</span> of a chemical is worth <span class="math">\((x^{1/r})\)</span></li>
</ul>
</dd></dl>

<dl class="data">
<dt id="fitness_scale">
<tt class="descname">fitness_scale</tt><a class="headerlink" href="#fitness_scale" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 1</em></p>
<p>The Normalized value of an organism&#8217;s fitness. All fitnesses are scaled
by a factor such that an &#8220;average&#8221; organism would have a fitness equal to
<a class="reference internal" href="#fitness_scale" title="fitness_scale"><tt class="xref py py-data docutils literal"><span class="pre">fitness_scale</span></tt></a>. See <a class="reference internal" href="sim/sim.html#fitness"><em>Fitness and Reproduction</em></a> for more details.</p>
</dd></dl>

<dl class="data">
<dt id="move_speed">
<tt class="descname">move_speed</tt><a class="headerlink" href="#move_speed" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 10</em></p>
<p>The number of units moved by an organism per movement instruction</p>
</dd></dl>

<dl class="data">
<dt id="max_moves">
<tt class="descname">max_moves</tt><a class="headerlink" href="#max_moves" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 2</em></p>
<p>The number of moves an organism will make in a row before its mode resets
to gather.</p>
</dd></dl>

<dl class="data">
<dt id="child_location">
<tt class="descname">child_location</tt><a class="headerlink" href="#child_location" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = random</em></p>
<p>Determines the location of an organisms offspring upon reproduction.
Options are:</p>
<ul class="simple">
<li>self - the child begins where the parent was</li>
<li>random - the child begins in a random location on the grid</li>
</ul>
</dd></dl>

<dl class="data">
<dt id="genome_fitness_cost">
<tt class="descname">genome_fitness_cost</tt><a class="headerlink" href="#genome_fitness_cost" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 0.005</em></p>
<p>The fitness cost per codon in the genome. This is relative to
<a class="reference internal" href="#fitness_scale" title="fitness_scale"><tt class="xref py py-data docutils literal"><span class="pre">fitness_scale</span></tt></a>, not to absolute fitness. See SIMULATION for more
details.</p>
</dd></dl>

<dl class="data">
<dt id="protein_cost">
<tt class="descname">protein_cost</tt><a class="headerlink" href="#protein_cost" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = True</em></p>
<p>Whether or not a protein requires a mole of each constituent chemical/codon.
If this is true, the organism will receive those chemicals back if building
the protein fails.</p>
</dd></dl>

</div>
<div class="section" id="genome">
<h3>Genome<a class="headerlink" href="#genome" title="Permalink to this headline">¶</a></h3>
<dl class="data">
<dt id="num_gather">
<tt class="descname">num_gather</tt><a class="headerlink" href="#num_gather" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 1</em></p>
<p>The number of codons that instruct an organism to switch to gather mode</p>
</dd></dl>

<dl class="data">
<dt id="num_move">
<tt class="descname">num_move</tt><a class="headerlink" href="#num_move" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 0</em></p>
<p>The number of codons that instruct an organism to switch to move mode</p>
</dd></dl>

<dl class="data">
<dt id="num_protein">
<tt class="descname">num_protein</tt><a class="headerlink" href="#num_protein" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 1</em></p>
<p>The number of codons that instruct an organism to switch to protein mode</p>
</dd></dl>

<dl class="data">
<dt id="copy_probability">
<tt class="descname">copy_probability</tt><a class="headerlink" href="#copy_probability" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = .01</em></p>
<p>The probability that a new organism will experience gene copying</p>
</dd></dl>

<dl class="data">
<dt id="delete_probability">
<tt class="descname">delete_probability</tt><a class="headerlink" href="#delete_probability" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = .01</em></p>
<p>The probability that a new organism will experience gene deletion</p>
</dd></dl>

<dl class="data">
<dt id="double_probability">
<tt class="descname">double_probability</tt><a class="headerlink" href="#double_probability" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = .01</em></p>
<p>The probability that a new organism will experience genome doubling</p>
</dd></dl>

<dl class="data">
<dt id="point_mutate_chance">
<tt class="descname">point_mutate_chance</tt><a class="headerlink" href="#point_mutate_chance" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = .002</em></p>
<p>The probability that any given codon in a new genome will mutate to
something else.</p>
</dd></dl>

<dl class="data">
<dt id="mode_switch_probability">
<tt class="descname">mode_switch_probability</tt><a class="headerlink" href="#mode_switch_probability" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 0.03</em></p>
<p>The probability that a point mutation yields a mode switch codon instead of a chemical codon.</p>
</dd></dl>

<dl class="data">
<dt id="swap_probability">
<tt class="descname">swap_probability</tt><a class="headerlink" href="#swap_probability" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = .01</em></p>
<p>The probability that an organism will copy a portion of another
organism&#8217;s genome.</p>
</dd></dl>

<dl class="data">
<dt id="initial_genome_length">
<tt class="descname">initial_genome_length</tt><a class="headerlink" href="#initial_genome_length" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 40</em></p>
<p>The starting genome length for all organisms.</p>
</dd></dl>

<dl class="data">
<dt id="initial_genome_creation">
<tt class="descname">initial_genome_creation</tt><a class="headerlink" href="#initial_genome_creation" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = random</em></p>
<p>The nature of the genomes of the first generation. Options are:</p>
<ul class="simple">
<li>random - The genome is comprised of random chemical codons (that don&#8217;t
require proteins to harvest) with a gather instruction at the beginning.</li>
<li>uniform - The genome is comprised entirely of chemical 0 codons, except
for the gather instruction at the beginning.</li>
</ul>
</dd></dl>

<dl class="data">
<dt id="min_genome_length">
<tt class="descname">min_genome_length</tt><a class="headerlink" href="#min_genome_length" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 20</em></p>
<p>The minimum allowed genome length. Small values make it more likely that
all organisms will converge on an extremely short gathering genome with
no real possiblity of protein creation.</p>
</dd></dl>

<dl class="data">
<dt id="genome_cap_ratio">
<tt class="descname">genome_cap_ratio</tt><a class="headerlink" href="#genome_cap_ratio" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 0.2</em></p>
<p>If non-zero, the length of an organism&#8217;s genome will be capped at
<a class="reference internal" href="#genome_cap_ratio" title="genome_cap_ratio"><tt class="xref py py-data docutils literal"><span class="pre">genome_cap_ratio</span></tt></a> * <a class="reference internal" href="#generation_length" title="generation_length"><tt class="xref py py-data docutils literal"><span class="pre">generation_length</span></tt></a></p>
</dd></dl>

<dl class="data">
<dt id="max_cut_del">
<tt class="descname">max_cut_del</tt><a class="headerlink" href="#max_cut_del" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 30</em></p>
<p>The maximum length of genome that can be copied, deleted, or transfered.</p>
</dd></dl>

</div>
<div class="section" id="protein">
<h3>Protein<a class="headerlink" href="#protein" title="Permalink to this headline">¶</a></h3>
<dl class="data">
<dt id="min_protein_length">
<tt class="descname">min_protein_length</tt><a class="headerlink" href="#min_protein_length" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 3</em></p>
<p>The minimum number of chemicals in a protein.</p>
</dd></dl>

<dl class="data">
<dt id="max_protein_length">
<tt class="descname">max_protein_length</tt><a class="headerlink" href="#max_protein_length" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 8</em></p>
<p>The maximum number of chemicals in a protein.</p>
</dd></dl>

<span class="target" id="strong-chance"></span><dl class="data">
<dt id="strong_chance">
<tt class="descname">strong_chance</tt><a class="headerlink" href="#strong_chance" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = .03</em></p>
<p>For each entry in the protein function vector, the probability that the
entry will have a &#8220;strong&#8221; function.</p>
</dd></dl>

<dl class="data">
<dt id="strong_factor">
<tt class="descname">strong_factor</tt><a class="headerlink" href="#strong_factor" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 10</em></p>
<p>The average value of a &#8220;strong&#8221; function element in a protein. For
complexes, the value is <span class="math">\(factor^{length}\)</span>.</p>
</dd></dl>

<dl class="data">
<dt id="weak_factor">
<tt class="descname">weak_factor</tt><a class="headerlink" href="#weak_factor" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = .01</em></p>
<p>The range of values for a &#8220;weak&#8221; function element in a protein. The
values can range from <span class="math">\(-factor*length\)</span> to <span class="math">\(factor*length\)</span>.</p>
</dd></dl>

<dl class="data">
<dt id="range_factor">
<tt class="descname">range_factor</tt><a class="headerlink" href="#range_factor" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 2.0</em></p>
<p>The variation in complex strengths. For each entry in the function vector,
a number N between -1 and 1 is chosen, and that entry is multiplied by
<span class="math">\(range\_factor^{N}\)</span>.</p>
</dd></dl>

<dl class="data">
<dt id="poison_chance">
<tt class="descname">poison_chance</tt><a class="headerlink" href="#poison_chance" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = .01</em></p>
<p>For each entry in the protein function vector, the probability that the
entry will have a negative &#8220;poisonous&#8221; function.</p>
</dd></dl>

<dl class="data">
<dt id="poison_factor">
<tt class="descname">poison_factor</tt><a class="headerlink" href="#poison_factor" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 5.0</em></p>
<p>The Average absolute value of a &#8220;poisonous&#8221; function element in a protein. For
complexes, the value is <span class="math">\(-factor^{length}\)</span>.</p>
</dd></dl>

</div>
<div class="section" id="complex">
<h3>Complex<a class="headerlink" href="#complex" title="Permalink to this headline">¶</a></h3>
<dl class="data">
<dt id="max_complex_length">
<tt class="descname">max_complex_length</tt><a class="headerlink" href="#max_complex_length" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 6</em></p>
<p>The maximum number of proteins in a complex.</p>
</dd></dl>

<dl class="data">
<dt id="max_binding_affinity">
<tt class="descname">max_binding_affinity</tt><a class="headerlink" href="#max_binding_affinity" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = .9</em></p>
<p>The maximum possible complex binding affinity.</p>
</dd></dl>

<dl class="data">
<dt id="min_binding_affinity">
<tt class="descname">min_binding_affinity</tt><a class="headerlink" href="#min_binding_affinity" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = .2</em></p>
<p>The minimum possible complex binding affinity.</p>
</dd></dl>

<dl class="data">
<dt id="maximum_bound">
<tt class="descname">maximum_bound</tt><a class="headerlink" href="#maximum_bound" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = .9</em></p>
<p>The amount of a complex that will bind. During each complex binding
reaction, no more than maximum_bound of the total amount will bind.</p>
</dd></dl>

<dl class="data">
<dt id="complex_complex_binding">
<tt class="descname">complex_complex_binding</tt><a class="headerlink" href="#complex_complex_binding" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = False</em></p>
<p>If True, complexes can bind with other (non-protein) complexes. If False,
only proteins can bind with (non-protein) complexes.</p>
</dd></dl>

</div>
<div class="section" id="family">
<h3>Family<a class="headerlink" href="#family" title="Permalink to this headline">¶</a></h3>
<dl class="data">
<dt id="family_binding_chance">
<tt class="descname">family_binding_chance</tt><a class="headerlink" href="#family_binding_chance" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 0.05</em></p>
<p>The chance that any two complex families will bind together.</p>
</dd></dl>

<dl class="data">
<dt id="sibling_distance">
<tt class="descname">sibling_distance</tt><a class="headerlink" href="#sibling_distance" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 1</em></p>
<p>The number of point differences a protein can have to another protein and
still be considered &#8220;siblings&#8221;.</p>
</dd></dl>

<dl class="data">
<dt id="max_relation_distance">
<tt class="descname">max_relation_distance</tt><a class="headerlink" href="#max_relation_distance" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 2</em></p>
<p>The number of point differences a protein can have to a family&#8217;s base
vector and still become part of that family. Effectively limits family
sizes.</p>
</dd></dl>

<dl class="data">
<dt id="separation_chance">
<tt class="descname">separation_chance</tt><a class="headerlink" href="#separation_chance" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = .05</em></p>
<p>The chance that a protein&#8211;even though it&#8217;s related to another
protein&#8211;will start a separate family.</p>
</dd></dl>

<dl class="data">
<dt id="mutate_factor">
<tt class="descname">mutate_factor</tt><a class="headerlink" href="#mutate_factor" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = 1.05</em></p>
<p>The amount of variation in function elements within a family.</p>
</dd></dl>

</div>
<div class="section" id="data">
<h3>Data<a class="headerlink" href="#data" title="Permalink to this headline">¶</a></h3>
<dl class="data">
<dt id="tracked_ancestors">
<tt class="descname">tracked_ancestors</tt><a class="headerlink" href="#tracked_ancestors" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = [1,3,10]</em></p>
<p>Specifies the numbers of generations back for which the ancestry data
should be logged.</p>
</dd></dl>

</div>
<div class="section" id="runtime">
<h3>Runtime<a class="headerlink" href="#runtime" title="Permalink to this headline">¶</a></h3>
<dl class="data">
<dt id="recorded_actions">
<tt class="descname">recorded_actions</tt><a class="headerlink" href="#recorded_actions" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = nothing</em></p>
<p>Recorded_actions is a list of all preset changes you want to make to the
simulation during runtime. Recorded_actions is not declared explicitly
here. It is constructed from a list of actions that you define here.
These actions can either change the value of one of these configuration
variables, or can call one of the predefined modifier functions: (no
functions yet).</p>
<p>For example, if you wanted to change the gather_proportion on generation
42, you would add:</p>
<p><tt class="docutils literal"><span class="pre">gen42</span> <span class="pre">=</span> <span class="pre">[&quot;gv.gather_proportion</span> <span class="pre">=</span> <span class="pre">.10&quot;]</span></tt></p>
<p>All of these commands are read by <a class="reference internal" href="utils/globals.html#module-pykaryote.utils.globals" title="pykaryote.utils.globals"><tt class="xref py py-mod docutils literal"><span class="pre">globals</span></tt></a> and
combined into recorded_actions.</p>
</dd></dl>

</div>
<div class="section" id="threshold-metrics">
<h3>threshold-metrics<a class="headerlink" href="#threshold-metrics" title="Permalink to this headline">¶</a></h3>
<p>This section is used only for data collection and graphing. It does not effect the results of a simulation.</p>
<p>This section allows one to record how many generations it takes a metric to reach a threshold. A list of metrics can be found in the source code of <a class="reference internal" href="utils/analysis.html#module-pykaryote.utils.analysis" title="pykaryote.utils.analysis"><tt class="xref py py-mod docutils literal"><span class="pre">analysis</span></tt></a>.</p>
<p>Options from this section are used to gather data for the gens_until_x bar graphs.</p>
<p>For example, to keep track of the number of generations it takes for complexity to reach 4.0, one would specify:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">metrics</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;complexity_avg&quot;</span><span class="p">]</span>
<span class="n">complexity_ave</span> <span class="o">=</span> <span class="p">[</span><span class="mf">4.0</span><span class="p">]</span>
</pre></div>
</div>
<dl class="data">
<dt id="metrics">
<tt class="descname">metrics</tt><a class="headerlink" href="#metrics" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = [&#8220;complexity_avg&#8221;, &#8220;genome_length_avg&#8221;]</em></p>
<p>The metrics to track. A list of metrics can be found int the source code of <a class="reference internal" href="utils/analysis.html#module-pykaryote.utils.analysis" title="pykaryote.utils.analysis"><tt class="xref py py-mod docutils literal"><span class="pre">analysis</span></tt></a>.</p>
</dd></dl>

<dl class="data">
<dt id="thresholds">
<tt class="descname">thresholds</tt><a class="headerlink" href="#thresholds" title="Permalink to this definition">¶</a></dt>
<dd><p><em>default = [[0.5, 1.0, 2.0, 3.0, 4.0, 4.5, 5.0, 5.5, 6.0], [50, 100, 150, 200, 250, 300]]</em></p>
<p>The thresholds to track for each metric. A list of lists of threshold values. Each list corresponds to a metric in <tt class="docutils literal"><span class="pre">metrics</span></tt>. Therefore, <cite>metrics</cite> and <cite>thresholds</cite> must have the same length.</p>
</dd></dl>

</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Simulation Configuration Files</a><ul>
<li><a class="reference internal" href="#configuration-options">Configuration Options</a><ul>
<li><a class="reference internal" href="#global">global</a></li>
<li><a class="reference internal" href="#environment">Environment</a></li>
<li><a class="reference internal" href="#organism">Organism</a></li>
<li><a class="reference internal" href="#genome">Genome</a></li>
<li><a class="reference internal" href="#protein">Protein</a></li>
<li><a class="reference internal" href="#complex">Complex</a></li>
<li><a class="reference internal" href="#family">Family</a></li>
<li><a class="reference internal" href="#data">Data</a></li>
<li><a class="reference internal" href="#runtime">Runtime</a></li>
<li><a class="reference internal" href="#threshold-metrics">threshold-metrics</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="genome.html"
                        title="previous chapter">The Model in Depth</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="data-format.html"
                        title="next chapter">Analyzer Data File Format</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/ref/conf.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="data-format.html" title="Analyzer Data File Format"
             >next</a> |</li>
        <li class="right" >
          <a href="genome.html" title="The Model in Depth"
             >previous</a> |</li>
        <li><a href="../index.html">Pykaryote 1.1.0d documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2011 - 2013, James Lamine and Ethan Van Andel.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b1.
    </div>
  </body>
</html>