<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>pydna Package &mdash; Pydna 0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="Pydna 0.1 documentation" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="index.html">Pydna 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="pydna-package">
<h1>pydna Package<a class="headerlink" href="#pydna-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-pydna.dsdna">
<span id="dsdna-module"></span><h2><tt class="xref py py-mod docutils literal"><span class="pre">dsdna</span></tt> Module<a class="headerlink" href="#module-pydna.dsdna" title="Permalink to this headline">¶</a></h2>
<p>Provides two classes, Dseq and Dseqrecord, for handling double stranded
DNA sequences. Dseq and Dseqrecord are subclasses of Biopythons
Seq and SeqRecord classes, respectively. These classes support the
notion of circular and linear DNA.</p>
<dl class="class">
<dt id="pydna.dsdna.Dseq">
<em class="property">class </em><tt class="descclassname">pydna.dsdna.</tt><tt class="descname">Dseq</tt><big>(</big><em>watson</em>, <em>crick=None</em>, <em>ovhg=None</em>, <em>linear=None</em>, <em>circular=None</em>, <em>alphabet=IUPACAmbiguousDNA()</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">Bio.Seq.Seq</span></tt></p>
<p>Dseq is a class designed to hold information for a double stranded
DNA fragment. Dseq also holds information describing the topology of
the DNA fragment (linear or circular).</p>
<p>Dseq is a subclass of the Biopython Seq object. It stores two
strings representing the watson (sense) and crick(antisense) strands.
two properties called linear and circular, and a numeric value ovhg
(overhang) describing the stagger for the watson and crick strand
in the 5&#8217; end of the fragment.</p>
<p>The most common usage is probably to create a Dseq object as a
part of a Dseqrecord object (see Dseqrecord).</p>
<p>There are three ways of creating a Dseq object directly:</p>
<p>Only one argument (string):</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">)</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
</pre></div>
</div>
<p>The given string will be interpreted as the watson strand of a
blunt, linear double stranded sequence object. The crick strand
is created automatically from the watson strand.</p>
<p>Two arguments (string, string):</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;gggaaat&quot;</span><span class="p">,</span><span class="s">&quot;ttt&quot;</span><span class="p">)</span>
<span class="go">Dseq(-7)</span>
<span class="go">gggaaat</span>
<span class="go">   ttt</span>
</pre></div>
</div>
<p>If both watson and crick are given, but not ovhg an attempt
will be made to find the best annealing between the strands.
There are limitations to this! For long fragments it is quite
slow. The length of the annealing sequences have to be at least
half the length of the shortest of the strands.</p>
<p>Three arguments (string, string, ovhg=int):</p>
<p>The ovhg parameter controls the stagger at the five prime end:</p>
<div class="highlight-python"><pre>ovhg=-2

XXXXX
  XXXXX

ovhg=-1

XXXXX
 XXXXX


ovhg=0

XXXXX
XXXXX

ovhg=1

 XXXXX
XXXXX

ovhg=2

  XXXXX
XXXXX</pre>
</div>
<p>Example of creating Dseq objects with different amounts of stagger:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-7)</span>
<span class="go">agt</span>
<span class="go">  attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Dseq(-6)</span>
<span class="go">agt</span>
<span class="go"> attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">agt</span>
<span class="go">attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go"> agt</span>
<span class="go">attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">  agt</span>
<span class="go">attca</span>
</pre></div>
</div>
<p>the ovhg parameter has to be given with both watson and crick,
otherwise an exception is raised.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s">&quot;agt&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
  File <span class="nb">&quot;/usr/local/lib/python2.7/dist-packages/pydna_/dsdna.py&quot;</span>, line <span class="m">169</span>, in <span class="n">__init__</span>
    <span class="k">else</span><span class="p">:</span>
<span class="gr">Exception</span>: <span class="n">ovhg defined without crick strand!</span>
</pre></div>
</div>
<p>The default alphabet is set to Biopython IUPACAmbiguousDNA</p>
<p>The shape of the fragment is set by either:</p>
<p>linear   = False, True</p>
<p>or</p>
<p>circular = True, False</p>
<p>Note that both ends of the DNA fragment has to be blunt to set
circular = True (or linear = False).</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span><span class="s">&quot;ttt&quot;</span><span class="p">)</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span><span class="s">&quot;ttt&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span> <span class="s">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">linear</span> <span class="o">=</span> <span class="bp">False</span> <span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(o3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span> <span class="s">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">circular</span> <span class="o">=</span> <span class="bp">True</span> <span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(o3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
</pre></div>
</div>
<p>Coercing to string</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;tttcccc&quot;</span><span class="p">,</span><span class="s">&quot;aaacccc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-11)</span>
<span class="go">    tttcccc</span>
<span class="go">ccccaaa</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
<span class="go">&#39;ggggtttcccc&#39;</span>
</pre></div>
</div>
<p>The double stranded part is accessible through the dsdata property:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">dsdata</span>
<span class="go">&#39;ttt&#39;</span>
</pre></div>
</div>
<p>Dseqrecord and Dseq share the same concept of length</p>
<div class="highlight-python"><pre>&lt;-- length --&gt;
GATCCTTT
     AAAGCCTAG</pre>
</div>
<p>The slicing of a linear Dseq object works mostly as it does for a string.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="s">&quot;ggatcc&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">&#39;a&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span>
<span class="go">&#39;at&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="go">&#39;&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[::</span><span class="mi">2</span><span class="p">]</span>
<span class="go">&#39;gac&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">linear</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">Dseq(-1)</span>
<span class="go">a</span>
<span class="go">t</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span>
<span class="go">Dseq(-2)</span>
<span class="go">at</span>
<span class="go">ta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="go">Dseq(-0)</span>


<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[::</span><span class="mi">2</span><span class="p">]</span>
<span class="go">Dseq(-3)</span>
<span class="go">gac</span>
<span class="go">ctg</span>
</pre></div>
</div>
<p>The slicing of a circular Dseq object has a slightly different meaning.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="s">&quot;ggAtCc&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span>
<span class="go">Dseq(o6)</span>
<span class="go">ggAtCc</span>
<span class="go">ccTaGg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">4</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">Dseq(-5)</span>
<span class="go">CcggA</span>
<span class="go">GgccT</span>
</pre></div>
</div>
<p>The slice [X:X] produces an empty slice for a string, while this will return
the linearized sequence starting at X:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="s">&quot;ggatcc&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span>
<span class="go">Dseq(o6)</span>
<span class="go">ggatcc</span>
<span class="go">cctagg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">3</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">Dseq(-6)</span>
<span class="go">tccgga</span>
<span class="go">aggcct</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
<dl class="method">
<dt id="pydna.dsdna.Dseq.T4">
<tt class="descname">T4</tt><big>(</big><em>nucleotides=None</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.T4"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.T4" title="Permalink to this definition">¶</a></dt>
<dd><p>Fill in of five prime protruding ends and chewing back of
three prime protruding ends by a DNA polymerase providing both
5&#8217;-3&#8217; DNA polymerase activity and 3&#8217;-5&#8217; nuclease acitivty
(such as T4 DNA polymerase). This in presence of any
combination of A, G, C or T. Default are all four
nucleotides together.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;gatcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatc</span>
<span class="go">ctagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">()</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatc</span>
<span class="go">ctagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s">&quot;t&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgat</span>
<span class="go"> tagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s">&quot;a&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcga</span>
<span class="go">  agctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s">&quot;g&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcg</span>
<span class="go">   gctag</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="pydna.dsdna.Dseq.circular">
<tt class="descname">circular</tt><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.circular"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.circular" title="Permalink to this definition">¶</a></dt>
<dd><p>The circular property</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.cut">
<tt class="descname">cut</tt><big>(</big><em>*enzymes</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.cut"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.cut" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of linear Dseq fragments produced in the digestion.
If there is not cut, the whole sequence is returned.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;ggatccnnngaattc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Dseq(-15)</span>
<span class="go">ggatccnnngaattc</span>
<span class="go">cctaggnnncttaag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Restriction</span> <span class="kn">import</span> <span class="n">BamHI</span><span class="p">,</span><span class="n">EcoRI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">type</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">))</span>
<span class="go">&lt;type &#39;list&#39;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">frag</span> <span class="ow">in</span> <span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="n">frag</span><span class="o">.</span><span class="n">fig</span><span class="p">()</span>
<span class="go">Dseq(-5)</span>
<span class="go">g</span>
<span class="go">cctag</span>
<span class="go">Dseq(-14)</span>
<span class="go">gatccnnngaattc</span>
<span class="go">    gnnncttaag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">EcoRI</span><span class="p">,</span> <span class="n">BamHI</span><span class="p">)</span> <span class="o">==</span>  <span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">,</span> <span class="n">EcoRI</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">c</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">EcoRI</span><span class="p">,</span> <span class="n">BamHI</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="o">+</span><span class="n">c</span>
<span class="go">Dseq(-15)</span>
<span class="go">ggatccnnngaattc</span>
<span class="go">cctaggnnncttaag</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.fig">
<tt class="descname">fig</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.fig"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.fig" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a representation of the sequence, truncated if
longer than 40 bp:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;atcgcttactagcgtactgatcatctgactgactagcgtga&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-41)</span>
<span class="go">atcg..gtga</span>
<span class="go">tagc..cact</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.fill_in">
<tt class="descname">fill_in</tt><big>(</big><em>nucleotides=None</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.fill_in"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.fill_in" title="Permalink to this definition">¶</a></dt>
<dd><p>Fill in of five prime protruding end with a DNA polymerase
that has only DNA polymerase activity (such as exo-klenow)
and any combination of A, G, C or T. Default are all four
nucleotides together.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span> <span class="s">&quot;ttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">fill_in</span><span class="p">()</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;caaa&quot;</span><span class="p">,</span> <span class="s">&quot;cttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaa</span>
<span class="go"> tttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">fill_in</span><span class="p">()</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaag</span>
<span class="go">gtttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">fill_in</span><span class="p">(</span><span class="s">&quot;g&quot;</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaag</span>
<span class="go">gtttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">fill_in</span><span class="p">(</span><span class="s">&quot;tac&quot;</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaa</span>
<span class="go"> tttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaac&quot;</span><span class="p">,</span> <span class="s">&quot;tttg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaac&quot;</span><span class="p">,</span> <span class="s">&quot;tttg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span>
<span class="go">Dseq(-5)</span>
<span class="go"> aaac</span>
<span class="go">gttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">fill_in</span><span class="p">()</span>
<span class="go">Dseq(-5)</span>
<span class="go"> aaac</span>
<span class="go">gttt</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.find">
<tt class="descname">find</tt><big>(</big><em>sub</em>, <em>start=0</em>, <em>end=9223372036854775807</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.find"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.find" title="Permalink to this definition">¶</a></dt>
<dd><p>Find method, like that of a python string.</p>
<p>This behaves like the python string method of the same name.</p>
<p>Returns an integer, the index of the first occurrence of substring
argument sub in the (sub)sequence given by [start:end].</p>
<dl class="docutils">
<dt>Arguments:</dt>
<dd><ul class="first last simple">
<li>sub - a string or another Seq object to look for</li>
<li>start - optional integer, slice start</li>
<li>end - optional integer, slice end</li>
</ul>
</dd>
</dl>
<p>Returns -1 if the subsequence is NOT found.</p>
<p>e.g. Locating the first typical start codon, AUG, in an RNA sequence:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s">&quot;GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&quot;AUG&quot;</span><span class="p">)</span>
<span class="go">3</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.five_prime_end">
<tt class="descname">five_prime_end</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.five_prime_end"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.five_prime_end" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a tuple describing the structure of the 5&#8217; end of
the DNA fragment</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span> <span class="s">&quot;ttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
<span class="go">(&#39;blunt&#39;, &#39;&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span> <span class="s">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go"> aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
<span class="go">(&quot;3&#39;&quot;, &#39;t&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span> <span class="s">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go">aaa</span>
<span class="go"> ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
<span class="go">(&quot;5&#39;&quot;, &#39;a&#39;)</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="pydna.dsdna.Dseq.linear">
<tt class="descname">linear</tt><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.linear"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.linear" title="Permalink to this definition">¶</a></dt>
<dd><p>The linear property</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.looped">
<tt class="descname">looped</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.looped"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.looped" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a circularized Dseq object. This can only be done if the
two ends are compatible, otherwise a TypeError is raised.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;catcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="go">Dseq(o8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s">&quot;t&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgat</span>
<span class="go"> tagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s">&quot;t&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="go">Dseq(o7)</span>
<span class="go">catcgat</span>
<span class="go">gtagcta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s">&quot;a&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcga</span>
<span class="go">  agctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s">&quot;a&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
  File <span class="nb">&quot;/usr/local/lib/python2.7/dist-packages/pydna/dsdna.py&quot;</span>, line <span class="m">357</span>, in <span class="n">looped</span>
    <span class="k">if</span> <span class="n">type5</span> <span class="o">==</span> <span class="n">type3</span> <span class="ow">and</span> <span class="nb">str</span><span class="p">(</span><span class="n">sticky5</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">rc</span><span class="p">(</span><span class="n">sticky3</span><span class="p">)):</span>
<span class="gr">TypeError</span>: <span class="n">DNA cannot be circularized.</span>
<span class="go">5&#39; and 3&#39; sticky ends not compatible!</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.mung">
<tt class="descname">mung</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.mung"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.mung" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulates treatment a nuclease with 5&#8217;-3&#8217; and 3&#8217;-5&#8217; single
strand specific exonuclease activity (such as mung bean nuclease):</p>
<div class="highlight-python"><pre>    ggatcc    -&gt;     gatcc
     ctaggg          ctagg

     ggatcc   -&gt;      ggatc
    tcctag            cctag

&gt;&gt;&gt; import pydna
&gt;&gt;&gt; b=pydna.Dseq("caaa", "cttt")
&gt;&gt;&gt; b
Dseq(-5)
caaa
 tttc
&gt;&gt;&gt; b.mung()
Dseq(-3)
aaa
ttt
&gt;&gt;&gt; c=pydna.Dseq("aaac", "tttg")
&gt;&gt;&gt; c
Dseq(-5)
 aaac
gttt
&gt;&gt;&gt; c.mung()
Dseq(-3)
aaa
ttt</pre>
</div>
</dd></dl>

<dl class="attribute">
<dt id="pydna.dsdna.Dseq.ovhg">
<tt class="descname">ovhg</tt><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.ovhg"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.ovhg" title="Permalink to this definition">¶</a></dt>
<dd><p>The ovhg property</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.rc">
<tt class="descname">rc</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.rc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.rc" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias of the reverse_complement method</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.reverse_complement">
<tt class="descname">reverse_complement</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.reverse_complement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a Dseq object where watson and crick have switched
places.</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.t4">
<tt class="descname">t4</tt><big>(</big><em>*args</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.t4"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.t4" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for the T4 method</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.three_prime_end">
<tt class="descname">three_prime_end</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.three_prime_end"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.three_prime_end" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a tuple describing the structure of the 5&#8217; end of
the DNA fragment</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span> <span class="s">&quot;ttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>
<span class="go">(&#39;blunt&#39;, &#39;&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span> <span class="s">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go"> aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>
<span class="go">(&quot;3&#39;&quot;, &#39;a&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span> <span class="s">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go">aaa</span>
<span class="go"> ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>
<span class="go">(&quot;5&#39;&quot;, &#39;t&#39;)</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.tolinear">
<tt class="descname">tolinear</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.tolinear"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.tolinear" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a blunt, linear copy of a circular Dseq object. This can
only be done if the Dseq object is circular, otherwise a
TypeError is raised.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;catcgatc&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(o8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">tolinear</span><span class="p">()</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pydna.dsdna.Dseqrecord">
<em class="property">class </em><tt class="descclassname">pydna.dsdna.</tt><tt class="descname">Dseqrecord</tt><big>(</big><em>record</em>, <em>circular=None</em>, <em>linear=None</em>, <em>*args</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">Bio.SeqRecord.SeqRecord</span></tt></p>
<p>Dseqrecord is a double stranded version of the Biopython SeqRecord class.
The Dseqrecord object holds a Dseq object describing the sequence.
Additionally, Dseqrecord hold meta information about the sequence in the
from of a list of SeqFeatures, in the same way as the SeqRecord does.
The Dseqrecord can be initialized with a string, Seq, Dseq, SeqRecord
or another Dseqrecord. The sequence information will be stored in a
Dseq object in all cases. Dseqrecord objects can be read or parsed
from sequences in Fasta, Embl or Genbank format.</p>
<p>There is a short representation associated with the Dseqrecord.
<tt class="docutils literal"><span class="pre">Dseqrecord(-3)</span></tt> represents a linear sequence of length 2
while <tt class="docutils literal"><span class="pre">Dseqrecord(o7)</span></tt>
represents a circular sequence of length 7.</p>
<p>Dseqrecord and Dseq share the same concept of length</p>
<div class="highlight-python"><pre>&lt;-- length --&gt;
GATCCTTT
     AAAGCCTAG</pre>
</div>
<dl class="docutils">
<dt>record <span class="classifier-delimiter">:</span> <span class="classifier">string, Seq, SeqRecord, Dseq or other Dseqrecord object</span></dt>
<dd>This data will be used to form the seq property</dd>
<dt>circular <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd>True or False reflecting the shape of the DNA molecule</dd>
<dt>linear <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd>True or False reflecting the shape of the DNA molecule</dd>
</dl>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">alphabet</span>
<span class="go">IUPACAmbiguousDNA()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">alphabet</span>
<span class="go">IUPACAmbiguousDNA()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">alphabet</span>
<span class="go">IUPACAmbiguousDNA()</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
<dl class="attribute">
<dt id="pydna.dsdna.Dseqrecord.circular">
<tt class="descname">circular</tt><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.circular"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.circular" title="Permalink to this definition">¶</a></dt>
<dd><p>Not really a method, but the circular property</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.cut">
<tt class="descname">cut</tt><big>(</big><em>*enzymes</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.cut"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.cut" title="Permalink to this definition">¶</a></dt>
<dd><p>Digest the Dseqrecord object with one or more restriction enzymes.
returns a list of linear Dseqrecords.</p>
<dl class="docutils">
<dt>enzymes <span class="classifier-delimiter">:</span> <span class="classifier">iterable object</span></dt>
<dd>iterable containing Biopython
restriction enzyme objects</dd>
</dl>
<dl class="docutils">
<dt>fragments <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd>list of Dseqrecord objects formed by the digestion</dd>
</dl>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;ggatcc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Restriction</span> <span class="kn">import</span> <span class="n">BamHI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">)</span>
<span class="go">[Dseqrecord(-5), Dseqrecord(-5)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frag1</span><span class="p">,</span> <span class="n">frag2</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frag1</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-5)</span>
<span class="go">g</span>
<span class="go">cctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frag2</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-5)</span>
<span class="go">gatcc</span>
<span class="go">    g</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.format">
<tt class="descname">format</tt><big>(</big><em>f='gb'</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.format"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.format" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the sequence as a string using a format supported by Biopython
SeqIO. Default is &#8220;gb&#8221; which is short for Genbank.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;02-FEB-2013&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">a</span><span class="o">.</span><span class="n">format</span><span class="p">()</span>
<span class="go">LOCUS       .                          3 bp    DNA     linear   UNK 02-FEB-2013</span>
<span class="go">DEFINITION  .</span>
<span class="go">ACCESSION   &lt;unknown id&gt;</span>
<span class="go">VERSION     &lt;unknown id&gt;</span>
<span class="go">KEYWORDS    .</span>
<span class="go">SOURCE      .</span>
<span class="go">  ORGANISM  .</span>
<span class="go">            .</span>
<span class="go">FEATURES             Location/Qualifiers</span>
<span class="go">ORIGIN</span>
<span class="go">        1 aaa</span>
<span class="go">//</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="pydna.dsdna.Dseqrecord.linear">
<tt class="descname">linear</tt><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.linear"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.linear" title="Permalink to this definition">¶</a></dt>
<dd><p>Not really a method, but the linear property</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.looped">
<tt class="descname">looped</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.looped"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.looped" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a circular version of the Dseqrecord object. The
underlying Dseq object has to have compatible ends.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(o3)</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.rc">
<tt class="descname">rc</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.rc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.rc" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of the reverse_complement method</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.reverse_complement">
<tt class="descname">reverse_complement</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.reverse_complement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a new Dseqrecord object which is the reverse complement.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.seguid">
<tt class="descname">seguid</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.seguid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.seguid" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the SEGUID for the sequence</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span>
<span class="go">&#39;YG7G6b2Kj/KtFOX63j8mRHHoIlE&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.shifted">
<tt class="descname">shifted</tt><big>(</big><em>shift</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.shifted"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.shifted" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a circular Dseqrecord with a new origin &lt;shift&gt;.
This only works on circular Dseqrecords. If we consider the following
circular sequence:</p>
<div class="line-block">
<div class="line"><tt class="docutils literal"><span class="pre">GAAAT</span>&nbsp;&nbsp; <span class="pre">&lt;--</span> <span class="pre">watson</span> <span class="pre">strand</span></tt></div>
<div class="line"><tt class="docutils literal"><span class="pre">CTTTA</span>&nbsp;&nbsp; <span class="pre">&lt;--</span> <span class="pre">crick</span> <span class="pre">strand</span></tt></div>
</div>
<p>The T and the G on the watson strand are linked together as well
as the A and the C of the of the crick strand.</p>
<p>if <tt class="docutils literal"><span class="pre">shift</span></tt> is 1, this indicates a new origin at position 1:</p>
<div class="line-block">
<div class="line-block">
<div class="line">new origin</div>
<div class="line"><br /></div>
</div>
<div class="line"><tt class="docutils literal"><span class="pre">G|AAAT</span></tt></div>
<div class="line"><tt class="docutils literal"><span class="pre">C|TTTA</span></tt></div>
</div>
<p>new sequence:</p>
<div class="line-block">
<div class="line"><tt class="docutils literal"><span class="pre">AAATG</span></tt></div>
<div class="line"><tt class="docutils literal"><span class="pre">TTTAC</span></tt></div>
</div>
<p>Shift is always positive and 0&lt;shift&lt;length, so in the example
below, permissible values of shift are 1,2 and 3</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;aaat&quot;</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(o4)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o4)</span>
<span class="go">aaat</span>
<span class="go">ttta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">shifted</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(o4)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o4)</span>
<span class="go">aata</span>
<span class="go">ttat</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.stamp">
<tt class="descname">stamp</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.stamp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.stamp" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a seguid stamp to the description property. This will
show in the genbank format. The following string:</p>
<p><tt class="docutils literal"><span class="pre">SEGUID</span> <span class="pre">&lt;seguid&gt;</span></tt></p>
<p>will be appended to the description property of the Dseqrecord
object (string).</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">stamp</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">description</span>
<span class="go">&#39;&lt;unknown description&gt; SEGUID YG7G6b2Kj/KtFOX63j8mRHHoIlE&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">verify_stamp</span><span class="p">()</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.synced">
<tt class="descname">synced</tt><big>(</big><em>ref</em>, <em>limit=25</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.synced"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.synced" title="Permalink to this definition">¶</a></dt>
<dd><p>This function returns a new circular sequence (Dseqrecord object), which has ben rotated
in such a way that there is maximum overlap between the sequence and
ref, which may be a string, Biopython Seq, SeqRecord object or
another Dseqrecord object.</p>
<p>The reason for using this could be to rotate a recombinant plasmid so
that it starts at the same position after cloning. See the example below:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;gaat&quot;</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o4)</span>
<span class="go">gaat</span>
<span class="go">ctta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">a</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span> <span class="o">+</span> <span class="n">a</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-4)</span>
<span class="go">atga</span>
<span class="go">tact</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">insert</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;CCC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recombinant</span> <span class="o">=</span> <span class="p">(</span><span class="n">d</span><span class="o">+</span><span class="n">insert</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recombinant</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o7)</span>
<span class="go">atgaCCC</span>
<span class="go">tactGGG</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recombinant</span><span class="o">.</span><span class="n">synced</span><span class="p">(</span><span class="n">a</span><span class="p">)</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o7)</span>
<span class="go">gaCCCat</span>
<span class="go">ctGGGta</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.tolinear">
<tt class="descname">tolinear</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.tolinear"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.tolinear" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a linear, blunt of a circular Dseqrecord object. The
underlying Dseq object has to be circular.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span> <span class="n">circular</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(o3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">tolinear</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(-3)</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.verify_stamp">
<tt class="descname">verify_stamp</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.verify_stamp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.verify_stamp" title="Permalink to this definition">¶</a></dt>
<dd><p>Verifies the SEGUID stamp in the description property is
valid. True if stamp match the sequid calculated from the sequence.
Exception raised if no stamp can be found.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;02-FEB-2013&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span>
<span class="go">&#39;YG7G6b2Kj/KtFOX63j8mRHHoIlE&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">a</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s">&quot;gb&quot;</span><span class="p">)</span>
<span class="go">LOCUS       .                          3 bp    DNA     linear   UNK 02-FEB-2013</span>
<span class="go">DEFINITION  .</span>
<span class="go">ACCESSION   &lt;unknown id&gt;</span>
<span class="go">VERSION     &lt;unknown id&gt;</span>
<span class="go">KEYWORDS    .</span>
<span class="go">SOURCE      .</span>
<span class="go">  ORGANISM  .</span>
<span class="go">            .</span>
<span class="go">FEATURES             Location/Qualifiers</span>
<span class="go">ORIGIN</span>
<span class="go">        1 aaa</span>
<span class="go">//</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">stamp</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">a</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s">&quot;gb&quot;</span><span class="p">)</span>
<span class="go">LOCUS       .                          3 bp    DNA     linear   UNK 02-FEB-2013</span>
<span class="go">DEFINITION  &lt;unknown description&gt; SEGUID YG7G6b2Kj/KtFOX63j8mRHHoIlE</span>
<span class="go">ACCESSION   &lt;unknown id&gt;</span>
<span class="go">VERSION     &lt;unknown id&gt;</span>
<span class="go">KEYWORDS    .</span>
<span class="go">SOURCE      .</span>
<span class="go">  ORGANISM  .</span>
<span class="go">            .</span>
<span class="go">FEATURES             Location/Qualifiers</span>
<span class="go">ORIGIN</span>
<span class="go">        1 aaa</span>
<span class="go">//</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">verify_stamp</span><span class="p">()</span>
<span class="go">True</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.write">
<tt class="descname">write</tt><big>(</big><em>filename=None</em>, <em>f='gb'</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.write"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.write" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes the Dseqrecord to a file using the format f, which must
be a format supported by Biopython SeqIO. Default is &#8220;gb&#8221;
which is short for Genbank.</p>
<p>Filename is the path to the file where the sequece is to be
written. The filename is optional, if it is not given, the
description property (string) is used together with the format:</p>
<p>If obj is the Dseqrecord object, the default file name will be:</p>
<p><tt class="docutils literal"><span class="pre">&lt;obj.description&gt;.&lt;f&gt;</span></tt></p>
<p>If the filename already exists and the sequence it contains
is different, a new file name will be used:</p>
<p><tt class="docutils literal"><span class="pre">&lt;obj.description&gt;_NEW.&lt;f&gt;</span></tt></p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="pydna.dsdna.parse">
<tt class="descclassname">pydna.dsdna.</tt><tt class="descname">parse</tt><big>(</big><em>data</em>, <em>filter=False</em>, <em>ds=True</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#parse"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.parse" title="Permalink to this definition">¶</a></dt>
<dd><p>This function returns <em>all</em> DNA sequences found in data. If no
sequences are found, an empty list is returned. This is a greedy
function, use carefully.</p>
<dl class="docutils">
<dt>data <span class="classifier-delimiter">:</span> <span class="classifier">string or iterable</span></dt>
<dd>see below</dd>
<dt>filter <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd>if filter == True, sequences will be silently filtered
for allowed characters (see docs for Dseqrecord)</dd>
<dt>ds <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd>Double stranded or single stranded DNA, Return
&#8220;Dseqrecord&#8221; or &#8220;SeqRecord&#8221; objects.</dd>
</dl>
<dl class="docutils">
<dt>list</dt>
<dd>contains Dseqrecord or SeqRecord objects</dd>
</dl>
<p>The data parameter is a string containing:</p>
<ol class="arabic simple">
<li>an absolute path to a local file.
The file will be read in text
mode and parsed for EMBL, FASTA
and Genbank sequences.</li>
<li>an absolute path to a local directory.
all files in the directory will be
read and parsed as in 1.</li>
<li>a string containing one or more
sequences in EMBL, GENBANK,
or FASTA format. Mixed formats
are allowed.</li>
<li>data can be a list or other iterable of 1 - 3</li>
</ol>
</dd></dl>

<dl class="function">
<dt id="pydna.dsdna.rc">
<tt class="descclassname">pydna.dsdna.</tt><tt class="descname">rc</tt><big>(</big><em>sequence</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#rc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.rc" title="Permalink to this definition">¶</a></dt>
<dd><p>returns the reverse complement of sequence (string)
accepts mixed DNA/RNA</p>
</dd></dl>

<dl class="function">
<dt id="pydna.dsdna.read">
<tt class="descclassname">pydna.dsdna.</tt><tt class="descname">read</tt><big>(</big><em>data</em>, <em>filter=False</em>, <em>ds=True</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#read"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.read" title="Permalink to this definition">¶</a></dt>
<dd><p>This function is similar the parse funtion but returns only
the first sequence found.</p>
<dl class="docutils">
<dt>data <span class="classifier-delimiter">:</span> <span class="classifier">string</span></dt>
<dd>see below</dd>
<dt>filter <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd>if filter == True, sequences will be silently filtered
for allowed characters (see docs for Dseqrecord)</dd>
<dt>ds <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd>Double stranded or single stranded DNA, Return
&#8220;Dseqrecord&#8221; or &#8220;SeqRecord&#8221; objects.</dd>
</dl>
<dl class="docutils">
<dt>Dseqrecord</dt>
<dd>contains the first Dseqrecord or SeqRecord object parsed.</dd>
</dl>
<p>The data parameter is similar to the data parameter for parse.</p>
<p>parse</p>
</dd></dl>

</div>
<div class="section" id="module-pydna.amplify">
<span id="amplify-module"></span><h2><tt class="xref py py-mod docutils literal"><span class="pre">amplify</span></tt> Module<a class="headerlink" href="#module-pydna.amplify" title="Permalink to this headline">¶</a></h2>
<p>This module provides functions for PCR.</p>
<p>Primers with 5&#8217; tails as well as inverse PCR on
circular templates are handled correctly.</p>
<dl class="class">
<dt id="pydna.amplify.Amplicon">
<em class="property">class </em><tt class="descclassname">pydna.amplify.</tt><tt class="descname">Amplicon</tt><big>(</big><em>forward_primer</em>, <em>reverse_primer</em>, <em>template</em>, <em>saltc=50</em>, <em>fc=1000</em>, <em>rc=1000</em><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#Amplicon"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.Amplicon" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">object</span></tt></p>
<p>Holds information about a PCR reaction involving two
primers and one template. This class is used by the
Anneal class and is not meant to be instantiated directly.</p>
<dl class="docutils">
<dt>forward_primer <span class="classifier-delimiter">:</span> <span class="classifier">SeqRecord(Biopython)</span></dt>
<dd>SeqRecord object holding the forward (sense) primer</dd>
<dt>reverse_primer <span class="classifier-delimiter">:</span> <span class="classifier">SeqRecord(Biopython)</span></dt>
<dd>SeqRecord object holding the reverse (antisense) primer</dd>
<dt>template <span class="classifier-delimiter">:</span> <span class="classifier">Dseqrecord</span></dt>
<dd>Dseqrecord object holding the template (circular or linear)</dd>
<dt>saltc <span class="classifier-delimiter">:</span> <span class="classifier">float, optional</span></dt>
<dd>saltc = monovalent cations (mM) (Na,K..)
default value is 50mM
This is used for Tm calculations.</dd>
<dt>fc <span class="classifier-delimiter">:</span> <span class="classifier">float, optional</span></dt>
<dd>primer concentration (nM)
default set to 1000nM = 1µM
This is used for Tm calculations.</dd>
<dt>rc <span class="classifier-delimiter">:</span> <span class="classifier">float, optional</span></dt>
<dd>primer concentration (nM)
default set to 1000nM = 1µM
This is used for Tm calculations.</dd>
</dl>
<dl class="method">
<dt id="pydna.amplify.Amplicon.detailed_figure">
<tt class="descname">detailed_figure</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#Amplicon.detailed_figure"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.Amplicon.detailed_figure" title="Permalink to this definition">¶</a></dt>
<dd><div class="highlight-python"><pre>5gctactacacacgtactgactg3
 |||||||||||||||||||||| tm 52.6 (dbd) 58.3
5gctactacacacgtactgactg...caagatagagtcagtaaccaca3
3cgatgatgtgtgcatgactgac...gttctatctcagtcattggtgt5
                          |||||||||||||||||||||| tm 49.1 (dbd) 57.7
                         3gttctatctcagtcattggtgt5</pre>
</div>
<dl class="docutils">
<dt>figure:string</dt>
<dd>A string containing a text representation of the primers
annealing on the template.</dd>
</dl>
<p>tm is the melting temperature (Tm) calculated according to
SantaLucia 1998 [1] for each primer.</p>
<p>(dbd) is Tm calculation for enzymes with dsDNA binding domains
like Pfu-Sso7d [2]. See [3] for more information.</p>
<table class="docutils footnote" frame="void" id="id1" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[1]</td><td><ol class="first last upperalpha simple" start="10">
<li>SantaLucia, “A Unified View of Polymer, Dumbbell, and Oligonucleotide DNA Nearest-neighbor Thermodynamics,” Proceedings of the National Academy of Sciences 95, no. 4 (1998): 1460.</li>
</ol>
</td></tr>
</tbody>
</table>
<table class="docutils footnote" frame="void" id="id2" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[2]</td><td><ol class="first last upperalpha simple" start="13">
<li>Nørholm, “A Mutant Pfu DNA Polymerase Designed for Advanced Uracil-excision DNA Engineering,” BMC Biotechnology 10, no. 1 (2010): 21, doi:10.1186/1472-6750-10-21.</li>
</ol>
</td></tr>
</tbody>
</table>
<table class="docutils footnote" frame="void" id="id3" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[3]</td><td><a class="reference external" href="http://www.thermoscientificbio.com/webtools/tmc/">http://www.thermoscientificbio.com/webtools/tmc/</a></td></tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pydna.amplify.Amplicon.flankdn">
<tt class="descname">flankdn</tt><big>(</big><em>flankdnlength=50</em><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#Amplicon.flankdn"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.Amplicon.flankdn" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a Dseqrecord object containing flankdnlength bases downstream of the reverse primer footprint.
Truncated if the template is not long enough.</p>
<div class="highlight-python"><pre>                               &lt;---- flankdn ------&gt;

                3actactgactatctTAATAA5
                 ||||||||||||||
acgcattcagctactgtactactgactatctatcgtacatgtactatcgtat</pre>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.amplify.Amplicon.flankup">
<tt class="descname">flankup</tt><big>(</big><em>flankuplength=50</em><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#Amplicon.flankup"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.Amplicon.flankup" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a Dseqrecord object containing flankuplength bases upstream of the forward primer footprint,
Truncated if the template is not long enough.</p>
<div class="highlight-python"><pre>&lt;--- flankup ---&gt;

          5TAATAAactactgactatct3
                 ||||||||||||||
acgcattcagctactgtactactgactatctatcg</pre>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.amplify.Amplicon.pcr_product">
<tt class="descname">pcr_product</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#Amplicon.pcr_product"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.Amplicon.pcr_product" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the PCR product as a Dseqrecord object.
Primers are marked as SeqFeatures(Biopython)
associated with the sequence.</p>
</dd></dl>

<dl class="method">
<dt id="pydna.amplify.Amplicon.pcr_program">
<tt class="descname">pcr_program</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#Amplicon.pcr_program"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.Amplicon.pcr_program" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a string containing a text representation of two proposed
PCR programs. The first program is adapted for Taq poymerase, while
the second is adapted for Pfu-Sso7d.</p>
<div class="highlight-python"><pre>Taq (rate 30 nt/s)
Three-step|         30 cycles     |      |SantaLucia 1998
94.0°C    |94.0°C                 |      |SaltC 50mM
__________|_____          72.0°C  |72.0°C|
04min00s  |30s  \         ________|______|
          |      \ 46.0°C/ 0min 1s|10min |
          |       \_____/         |      |
          |         30s           |      |4-8°C

Pfu-Sso7d (rate 15s/kb)
Three-step|          30 cycles   |      |Breslauer1986,SantaLucia1998
98.0°C    |98.0°C                |      |SaltC 50mM
__________|_____          72.0°C |72.0°C|Primer1C   1µM
00min30s  |10s  \ 61.0°C ________|______|Primer2C   1µM
          |      \______/ 0min 0s|10min |
          |        10s           |      |4-8°C</pre>
</div>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pydna.amplify.Anneal">
<em class="property">class </em><tt class="descclassname">pydna.amplify.</tt><tt class="descname">Anneal</tt><big>(</big><em>primers</em>, <em>template</em>, <em>homology_limit=13</em>, <em>max_product_size=15000</em><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#Anneal"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.Anneal" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">object</span></tt></p>
<dl class="docutils">
<dt>primers <span class="classifier-delimiter">:</span> <span class="classifier">iterable containing SeqRecord objects</span></dt>
<dd>Primer sequences 5&#8217;-3&#8217;.</dd>
<dt>template <span class="classifier-delimiter">:</span> <span class="classifier">Dseqrecord object</span></dt>
<dd>The template sequence 5&#8217;-3&#8217;.</dd>
<dt>homology_limit <span class="classifier-delimiter">:</span> <span class="classifier">int, optional</span></dt>
<dd>limit length of the annealing part of the primers.</dd>
<dt>max_product_size: int, optional</dt>
<dd>PCR products longer than max_product_size are not reported</dd>
</dl>
<dl class="docutils">
<dt>amplicons: list</dt>
<dd>A list of Amplicon objects, one for each primer pair that may form a PCR product.</dd>
</dl>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">template</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;tacactcaccgtctatcattatctactatcgactgtatcatctgatagcac&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p1</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&quot;&gt;p1</span><span class="se">\n</span><span class="s">tacactcaccgtctatcattatc&quot;</span><span class="p">,</span> <span class="n">ds</span> <span class="o">=</span> <span class="bp">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&quot;&gt;p2</span><span class="se">\n</span><span class="s">cgactgtatcatctgatagcac&quot;</span><span class="p">,</span> <span class="n">ds</span> <span class="o">=</span> <span class="bp">False</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Anneal</span><span class="p">((</span><span class="n">p1</span><span class="p">,</span><span class="n">p2</span><span class="p">),</span> <span class="n">template</span><span class="p">)</span>
<span class="go">Anneal(amplicons = 1)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Anneal</span><span class="p">((</span><span class="n">p1</span><span class="p">,</span><span class="n">p2</span><span class="p">),</span> <span class="n">template</span><span class="p">)</span><span class="o">.</span><span class="n">amplicons</span>
<span class="go">[Amplicon(51)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">amplicon_list</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">Anneal</span><span class="p">((</span><span class="n">p1</span><span class="p">,</span><span class="n">p2</span><span class="p">),</span> <span class="n">template</span><span class="p">)</span><span class="o">.</span><span class="n">amplicons</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">amplicon</span> <span class="o">=</span> <span class="n">amplicon_list</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">amplicon</span>
<span class="go">Amplicon(51)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">amplicon</span><span class="o">.</span><span class="n">detailed_figure</span><span class="p">()</span>
<span class="go">5tacactcaccgtctatcattatc...cgactgtatcatctgatagcac3</span>
<span class="go">                           |||||||||||||||||||||| tm 50.6 (dbd) 60.5</span>
<span class="go">                          3gctgacatagtagactatcgtg5</span>
<span class="go">5tacactcaccgtctatcattatc3</span>
<span class="go"> ||||||||||||||||||||||| tm 49.4 (dbd) 58.8</span>
<span class="go">3atgtgagtggcagatagtaatag...gctgacatagtagactatcgtg5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prod</span> <span class="o">=</span> <span class="n">amplicon</span><span class="o">.</span><span class="n">pcr_product</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prod</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;02-FEB-2013&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">prod</span>
<span class="go">Dseqrecord</span>
<span class="go">circular: False</span>
<span class="go">size: 51</span>
<span class="go">ID: 51bp U96+TO06Y6pFs74SQx8M1IVTBiY</span>
<span class="go">Name: 51bp_PCR_prod</span>
<span class="go">Description: Primers p1 &lt;unknown name&gt;</span>
<span class="go">Number of features: 2</span>
<span class="go">/date=02-FEB-2013</span>
<span class="go">Dseq(-51)</span>
<span class="go">taca..gcac</span>
<span class="go">atgt..cgtg</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
<dl class="method">
<dt id="pydna.amplify.Anneal.featured_template">
<tt class="descname">featured_template</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#Anneal.featured_template"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.Anneal.featured_template" title="Permalink to this definition">¶</a></dt>
<dd><p>returns the template Dseqrecord object with the footprint of all
annealing primers marked as SeqFeatures.</p>
</dd></dl>

<dl class="method">
<dt id="pydna.amplify.Anneal.products">
<tt class="descname">products</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#Anneal.products"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.Anneal.products" title="Permalink to this definition">¶</a></dt>
<dd><p>returns all pcr product as a list of Dseqrecords</p>
</dd></dl>

<dl class="method">
<dt id="pydna.amplify.Anneal.report">
<tt class="descname">report</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#Anneal.report"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.Anneal.report" title="Permalink to this definition">¶</a></dt>
<dd><p>This method is an alias of str(Annealobj).
Returns a short string representation.</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="pydna.amplify.basictm">
<tt class="descclassname">pydna.amplify.</tt><tt class="descname">basictm</tt><big>(</big><em>primer</em>, <em>*args</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#basictm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.basictm" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the melting temperature (Tm) of the primer using
the basic formula.</p>
<div class="line-block">
<div class="line">Tm = (wA+xT)*2 + (yG+zC)*4 assumed 50mM monovalent cations</div>
<div class="line"><br /></div>
<div class="line">w = number of A in primer</div>
<div class="line">x = number of T in primer</div>
<div class="line">y = number of G in primer</div>
<div class="line">z = number of C in primer</div>
</div>
<dl class="docutils">
<dt>primer <span class="classifier-delimiter">:</span> <span class="classifier">string</span></dt>
<dd>Primer sequence 5&#8217;-3&#8217;</dd>
</dl>
<p>tm : int</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.amplify</span> <span class="kn">import</span> <span class="n">basictm</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">basictm</span><span class="p">(</span><span class="s">&quot;ggatcc&quot;</span><span class="p">)</span>
<span class="go">20</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pydna.amplify.pcr">
<tt class="descclassname">pydna.amplify.</tt><tt class="descname">pcr</tt><big>(</big><em>*args</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#pcr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.pcr" title="Permalink to this definition">¶</a></dt>
<dd><p>pcr is a convenience function for Anneal to simplify its usage,
especially from the command line. If more than one PCR product is
formed, an exception is raised.</p>
<p>args is any iterable of sequences or an iterable of iterables of sequences.
args will be greedily flattened.</p>
<dl class="docutils">
<dt>args <span class="classifier-delimiter">:</span> <span class="classifier">iterable containing sequence objects</span></dt>
<dd>Several arguments are also accepted.</dd>
<dt>limit <span class="classifier-delimiter">:</span> <span class="classifier">int = 13, optional</span></dt>
<dd>limit length of the annealing part of the primers.</dd>
</dl>
<p>sequences in args could be of type:</p>
<p>string
Seq
SeqRecord
Dseqrecord</p>
<p>The last sequence will be interpreted as the template
all preceeding sequences as primers.</p>
<p>This is a powerful function, use with care!</p>
<dl class="docutils">
<dt>product <span class="classifier-delimiter">:</span> <span class="classifier">Dseqrecord</span></dt>
<dd>a Dseqrecord object representing the PCR product.
The direction of the PCR product will be the same as
for the template sequence.</dd>
</dl>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">template</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;tacactcaccgtctatcattatctactatcgactgtatcatctgatagcac&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p1</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&quot;&gt;p1</span><span class="se">\n</span><span class="s">tacactcaccgtctatcattatc&quot;</span><span class="p">,</span> <span class="n">ds</span> <span class="o">=</span> <span class="bp">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&quot;&gt;p2</span><span class="se">\n</span><span class="s">cgactgtatcatctgatagcac&quot;</span><span class="p">,</span> <span class="n">ds</span> <span class="o">=</span> <span class="bp">False</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">pcr</span><span class="p">(</span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">,</span> <span class="n">template</span><span class="p">)</span>
<span class="go">Dseqrecord(-51)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">pcr</span><span class="p">([</span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">],</span> <span class="n">template</span><span class="p">)</span>
<span class="go">Dseqrecord(-51)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">pcr</span><span class="p">((</span><span class="n">p1</span><span class="p">,</span><span class="n">p2</span><span class="p">,),</span> <span class="n">template</span><span class="p">)</span>
<span class="go">Dseqrecord(-51)</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pydna.amplify.tmbreslauer86">
<tt class="descclassname">pydna.amplify.</tt><tt class="descname">tmbreslauer86</tt><big>(</big><em>primer</em>, <em>dnac=500.0</em>, <em>saltc=50</em>, <em>thermodynamics=False</em><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#tmbreslauer86"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.tmbreslauer86" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the melting temperature (Tm) of the primer using
the nearest neighbour algorithm. Formula and thermodynamic data
is taken from Breslauer 1986. These data are no longer widely used.</p>
<p>Breslauer 1986, table 2 [1]</p>
<table border="1" class="docutils">
<colgroup>
<col width="29%" />
<col width="29%" />
<col width="24%" />
<col width="18%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">pair</th>
<th class="head">dH</th>
<th class="head">dS</th>
<th class="head">dG</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>AA/TT</td>
<td>9.1</td>
<td>24.0</td>
<td>1.9</td>
</tr>
<tr class="row-odd"><td>AT/TA</td>
<td>8.6</td>
<td>23.9</td>
<td>1.5</td>
</tr>
<tr class="row-even"><td>TA/AT</td>
<td>6.0</td>
<td>16.9</td>
<td>0.9</td>
</tr>
<tr class="row-odd"><td>CA/GT</td>
<td>5.8</td>
<td>12.9</td>
<td>1.9</td>
</tr>
<tr class="row-even"><td>GT/CA</td>
<td>6.5</td>
<td>17.3</td>
<td>1.3</td>
</tr>
<tr class="row-odd"><td>CT/GA</td>
<td>7.8</td>
<td>20.8</td>
<td>1.6</td>
</tr>
<tr class="row-even"><td>GA/CT</td>
<td>5.6</td>
<td>13.5</td>
<td>1.6</td>
</tr>
<tr class="row-odd"><td>CG/GC</td>
<td>11.9</td>
<td>27.8</td>
<td>3.6</td>
</tr>
<tr class="row-even"><td>GC/CG</td>
<td>11.1</td>
<td>26.7</td>
<td>3.1</td>
</tr>
<tr class="row-odd"><td>GG/CC</td>
<td>11.0</td>
<td>26.6</td>
<td>3.1</td>
</tr>
</tbody>
</table>
<dl class="docutils">
<dt>primer <span class="classifier-delimiter">:</span> <span class="classifier">string</span></dt>
<dd>Primer sequence 5&#8217;-3&#8217; in UPPERCASE</dd>
</dl>
<p>tm : float</p>
<table class="docutils footnote" frame="void" id="id4" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[1]</td><td>K.J. Breslauer et al., “Predicting DNA Duplex Stability from the Base Sequence,” Proceedings of the National Academy of Sciences 83, no. 11 (1986): 3746.</td></tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.amplify</span> <span class="kn">import</span> <span class="n">tmbreslauer86</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tmbreslauer86</span><span class="p">(</span><span class="s">&quot;ACGTCATCGACACTATCATCGAC&quot;</span><span class="p">)</span>
<span class="go">64.28863985851899</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pydna.amplify.tmbresluc">
<tt class="descclassname">pydna.amplify.</tt><tt class="descname">tmbresluc</tt><big>(</big><em>primer</em>, <em>primerc=500.0</em>, <em>saltc=50</em>, <em>thermodynamics=False</em><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#tmbresluc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.tmbresluc" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the tm for a primer using a formula adapted to polymerases
with a DNA binding domain.</p>
<dl class="docutils">
<dt>primer <span class="classifier-delimiter">:</span> <span class="classifier">string</span></dt>
<dd>primer sequence 5&#8217;-3&#8217;</dd>
<dt>primerc <span class="classifier-delimiter">:</span> <span class="classifier">float</span></dt>
<dd>concentration (nM)</dd>
<dt>saltc <span class="classifier-delimiter">:</span> <span class="classifier">float, optional</span></dt>
<dd>Monovalent cation concentration (mM)</dd>
<dt>thermodynamics <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd>prints details of the thermodynamic data to stdout. For
debugging only.</dd>
</dl>
<dl class="docutils">
<dt>tm <span class="classifier-delimiter">:</span> <span class="classifier">float</span></dt>
<dd>the tm of the primer</dd>
<dt>tm,dH,dS <span class="classifier-delimiter">:</span> <span class="classifier">tuple</span></dt>
<dd>tm and dH and dS used for the calculation</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pydna.amplify.tmstaluc98">
<tt class="descclassname">pydna.amplify.</tt><tt class="descname">tmstaluc98</tt><big>(</big><em>primer</em>, <em>dnac=50</em>, <em>saltc=50</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#tmstaluc98"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.tmstaluc98" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the melting temperature (Tm) of the primer using
the nearest neighbour algorithm. Formula and thermodynamic data
is taken from SantaLucia 1998. This implementation gives the same
answer as the one provided by Biopython (See Examples).</p>
<p>Thermodynamic data used:</p>
<table border="1" class="docutils">
<colgroup>
<col width="38%" />
<col width="31%" />
<col width="31%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">pair</th>
<th class="head">dH</th>
<th class="head">dS</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>AA/TT</td>
<td>7.9</td>
<td>22.2</td>
</tr>
<tr class="row-odd"><td>AT/TA</td>
<td>7.2</td>
<td>20.4</td>
</tr>
<tr class="row-even"><td>TA/AT</td>
<td>7.2</td>
<td>21.3</td>
</tr>
<tr class="row-odd"><td>CA/GT</td>
<td>8.5</td>
<td>22.7</td>
</tr>
<tr class="row-even"><td>GT/CA</td>
<td>8.4</td>
<td>22.4</td>
</tr>
<tr class="row-odd"><td>CT/GA</td>
<td>7.8</td>
<td>21.0</td>
</tr>
<tr class="row-even"><td>GA/CT</td>
<td>8.2</td>
<td>22.2</td>
</tr>
<tr class="row-odd"><td>CG/GC</td>
<td>10.6</td>
<td>27.2</td>
</tr>
<tr class="row-even"><td>GC/CG</td>
<td>9.8</td>
<td>24.4</td>
</tr>
<tr class="row-odd"><td>GG/CC</td>
<td>8.0</td>
<td>19.9</td>
</tr>
</tbody>
</table>
<dl class="docutils">
<dt>primer <span class="classifier-delimiter">:</span> <span class="classifier">string</span></dt>
<dd>Primer sequence 5&#8217;-3&#8217; in UPPERCASE</dd>
</dl>
<dl class="docutils">
<dt>tm <span class="classifier-delimiter">:</span> <span class="classifier">float</span></dt>
<dd>tm of the primer</dd>
</dl>
<table class="docutils footnote" frame="void" id="id5" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[1]</td><td><ol class="first last upperalpha simple" start="10">
<li>SantaLucia, “A Unified View of Polymer, Dumbbell, and Oligonucleotide DNA Nearest-neighbor Thermodynamics,” Proceedings of the National Academy of Sciences 95, no. 4 (1998): 1460.</li>
</ol>
</td></tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.amplify</span> <span class="kn">import</span> <span class="n">tmstaluc98</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqUtils.MeltingTemp</span> <span class="kn">import</span> <span class="n">Tm_staluc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tmstaluc98</span><span class="p">(</span><span class="s">&quot;ACGTCATCGACACTATCATCGAC&quot;</span><span class="p">)</span>
<span class="go">54.55597724052518</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Tm_staluc</span><span class="p">(</span><span class="s">&quot;ACGTCATCGACACTATCATCGAC&quot;</span><span class="p">)</span>
<span class="go">54.555977240525124</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="module-pydna.primer_design">
<span id="primer-design-module"></span><h2><tt class="xref py py-mod docutils literal"><span class="pre">primer_design</span></tt> Module<a class="headerlink" href="#module-pydna.primer_design" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pydna.primer_design.cloning_primers">
<tt class="descclassname">pydna.primer_design.</tt><tt class="descname">cloning_primers</tt><big>(</big><em>template</em>, <em>minlength=16</em>, <em>maxlength=29</em>, <em>fp=None</em>, <em>rp=None</em>, <em>fp_tail=''</em>, <em>rp_tail=''</em>, <em>target_tm=55.0</em>, <em>primerc=1000.0</em>, <em>saltc=50.0</em>, <em>formula='tmbresluc'</em><big>)</big><a class="reference internal" href="_modules/pydna/primer_design.html#cloning_primers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.primer_design.cloning_primers" title="Permalink to this definition">¶</a></dt>
<dd><p>This function can design one or two primers for PCR amplification of a 
given sequence. This function accepts a Dseqrecord object containing the 
template sequence and returns a pydna amplicon object.</p>
<p>The amplicon object contains the primers, a figure describing the how the 
primers anneal and two suggested PCR programmes.</p>
<dl class="docutils">
<dt>template <span class="classifier-delimiter">:</span> <span class="classifier">Dseqrecord</span></dt>
<dd>a Dseqrecord object.</dd>
<dt>minlength <span class="classifier-delimiter">:</span> <span class="classifier">int, optional</span></dt>
<dd>Minimum length of the annealing part of the primer</dd>
<dt>maxlength <span class="classifier-delimiter">:</span> <span class="classifier">int, optional</span></dt>
<dd>Maximum length (including tail) for designed primers.</dd>
<dt>fp, rp <span class="classifier-delimiter">:</span> <span class="classifier">SeqRecord, optional</span></dt>
<dd>optional Biopython SeqRecord objects containing one primer each.</dd>
<dt>fp_tail, rp_tail <span class="classifier-delimiter">:</span> <span class="classifier">string, optional</span></dt>
<dd>optional tails to be added to the forwars or reverse primers</dd>
<dt>target_tm <span class="classifier-delimiter">:</span> <span class="classifier">float, optional</span></dt>
<dd>target tm for the primers</dd>
<dt>primerc, saltc <span class="classifier-delimiter">:</span> <span class="classifier">float, optional</span></dt>
<dd>limit is set to 25 by default.</dd>
<dt>formula <span class="classifier-delimiter">:</span> <span class="classifier">string</span></dt>
<dd><p class="first">formula used for tm calculation
this is the name of a function.
built in options are:</p>
<p class="last">tmbresluc
basictm
tmstaluc98 
tmbreslauer86</p>
</dd>
</dl>
<p>These functions are imported from the pydna.amplify module, but can be 
substituted for some other custom mad function.</p>
<dl class="docutils">
<dt>frecs, cp <span class="classifier-delimiter">:</span> <span class="classifier">tuple</span></dt>
<dd><p class="first">frecs are the same Dseqrecords as given as arguments, but with the
regions of homology added to the features.</p>
<p class="last">cp is a list of Dseqrecords representing the circular products
sorted by length (long -&gt; short).</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-pydna.assembly">
<span id="assembly-module"></span><h2><tt class="xref py py-mod docutils literal"><span class="pre">assembly</span></tt> Module<a class="headerlink" href="#module-pydna.assembly" title="Permalink to this headline">¶</a></h2>
<p>Provides functions for assembly of sequences by homologous recombination.
Given a list of sequences (Dseqrecords), all sequences will be analyzed for
overlapping regions of DNA (common substrings).</p>
<p>The assembly algorithm is based on forming a network where each
overlapping sequence forms a node and intervening sequences form edges.</p>
<p>Then all possible linear or circular assemblies will be returned in the
order of their length.</p>
<dl class="function">
<dt id="pydna.assembly.circular_assembly">
<tt class="descclassname">pydna.assembly.</tt><tt class="descname">circular_assembly</tt><big>(</big><em>form_rec_list</em>, <em>limit=25</em><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#circular_assembly"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.circular_assembly" title="Permalink to this definition">¶</a></dt>
<dd><p>Accepts a list of Dseqrecords and tries to assemble them into a
circular assembly by homologous recombination based on shared
regions of homology with a minimum length given by limit.</p>
<dl class="docutils">
<dt>form_rec_list <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd>a list of Dseqrecord objects.</dd>
<dt>limit <span class="classifier-delimiter">:</span> <span class="classifier">int, optional</span></dt>
<dd>limit is set to 25 by default.</dd>
</dl>
<dl class="docutils">
<dt>frecs, cp <span class="classifier-delimiter">:</span> <span class="classifier">tuple</span></dt>
<dd><p class="first">frecs are the same Dseqrecords as given as arguments, but with the
regions of homology added to the features.</p>
<p class="last">cp is a list of Dseqrecords representing the circular products
sorted by length (long -&gt; short).</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pydna.assembly.linear_assembly">
<tt class="descclassname">pydna.assembly.</tt><tt class="descname">linear_assembly</tt><big>(</big><em>form_rec_list</em>, <em>limit=25</em><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#linear_assembly"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.linear_assembly" title="Permalink to this definition">¶</a></dt>
<dd><p>Accepts a list of Dseqrecords and tries to assemble them into a
linear assembly by homologous recombination based on shared
regions of homology with a minimum length given by limit.</p>
<dl class="docutils">
<dt>form_rec_list <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd>a list of Dseqrecord objects.</dd>
<dt>limit <span class="classifier-delimiter">:</span> <span class="classifier">int, optional</span></dt>
<dd>limit is set to 25 by default.</dd>
</dl>
<dl class="docutils">
<dt>frecs, lp <span class="classifier-delimiter">:</span> <span class="classifier">tuple</span></dt>
<dd><p class="first">frecs are the same Dseqrecords as given as arguments, but with the
regions of homology added to the features.</p>
<p class="last">lp is a list of Dseqrecords representing the linear products
sorted by length (long -&gt; short).</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-pydna.assembly2">
<span id="assembly2-module"></span><h2><tt class="xref py py-mod docutils literal"><span class="pre">assembly2</span></tt> Module<a class="headerlink" href="#module-pydna.assembly2" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pydna.assembly2.fusion_pcr_assembly">
<tt class="descclassname">pydna.assembly2.</tt><tt class="descname">fusion_pcr_assembly</tt><big>(</big><em>dsrecs</em>, <em>limit=15</em><big>)</big><a class="reference internal" href="_modules/pydna/assembly2.html#fusion_pcr_assembly"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly2.fusion_pcr_assembly" title="Permalink to this definition">¶</a></dt>
<dd><p>Accepts a list of Dseqrecords and tries to assemble them into
linear assemblies based on shared regions of homology at the 
extremities of each fragment with a minimum length given by limit.</p>
<dl class="docutils">
<dt>dsrecs <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd>a list of Dseqrecord objects.</dd>
<dt>limit <span class="classifier-delimiter">:</span> <span class="classifier">int, optional</span></dt>
<dd>limit is set to 15 nucleotides by default.</dd>
</dl>
<dl class="docutils">
<dt>linear: list</dt>
<dd><p class="first">Each object in linear is a named tuple with the following
fields:</p>
<table border="1" class="docutils">
<colgroup>
<col width="27%" />
<col width="18%" />
<col width="12%" />
<col width="42%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Field</th>
<th class="head" colspan="2">Type</th>
<th class="head">Contains</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>result</td>
<td colspan="2">Dseqrecord</td>
<td>Assembeled sequence</td>
</tr>
<tr class="row-odd"><td>fragments</td>
<td colspan="2">list of Dseqrecords</td>
<td>list of fragments</td>
</tr>
<tr class="row-even"><td>offsets</td>
<td colspan="2">list of integers</td>
<td>list of cumulative offsets
aligning each source
fragment</td>
</tr>
<tr class="row-odd"><td>overlap_sizes</td>
<td colspan="2">list of integers</td>
<td>list of the length of each
overlap joining the
sequences</td>
</tr>
</tbody>
</table>
<p>Source fragments are the double stranded blunt DNA fragments 
originally added as the dsrecs argument. The parwise overlaps 
found are added ar features to the source fragments.</p>
<p>Offsets is a list with the cumulative stagger between 
each fragment.</p>
<div class="last highlight-python"><pre>            &lt;--6-&gt;        &lt;--6-&gt;         overlap_sizes = [6,6]
                        
cggcggcgggccTGCCTC                      \
gccgccgcccggACGGAG                      | 
                                        |
            TGCCTCaccattgcAAAAAA        | fragments
            ACGGAGtggtaacgTTTTTT        |
                                        | 
                          AAAAAAcatcata | 
                          TTTTTTgtagtat /

&lt;----------&gt;
&lt;-----------------------&gt;               offsets = [12,26]</pre>
</div>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pydna.assembly2.gibson_assembly">
<tt class="descclassname">pydna.assembly2.</tt><tt class="descname">gibson_assembly</tt><big>(</big><em>dsrecs</em>, <em>limit=15</em><big>)</big><a class="reference internal" href="_modules/pydna/assembly2.html#gibson_assembly"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly2.gibson_assembly" title="Permalink to this definition">¶</a></dt>
<dd><p>Accepts a list of Dseqrecords and tries to assemble them into
linear or circular assemblies based on shared regions of homology at the 
extremities of each fragment with a minimum length given by limit.</p>
<p>This is a simulation of the DNA assembly method described in:</p>
<p>Gibson DG, Young L, Chuang R-Y, Venter JC, Hutchison CA 3rd, Smith HO (2009)
Enzymatic assembly of DNA molecules up to several hundred kilobases. 
Nat Methods 6:343–345. doi: 10.1038/nmeth.1318</p>
<dl class="docutils">
<dt>dsrecs <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd>a list of Dseqrecord objects.</dd>
<dt>limit <span class="classifier-delimiter">:</span> <span class="classifier">int, optional</span></dt>
<dd>limit is set to 15 nucleotides by default.</dd>
</dl>
<dl class="docutils">
<dt>linear, circular <span class="classifier-delimiter">:</span> <span class="classifier">tuple</span></dt>
<dd><p class="first">Linear and circular are two lists containing linear 
and circular assemblies.</p>
<p>Each object in linear or circular is a named tuple with the following
fields:</p>
<table border="1" class="docutils">
<colgroup>
<col width="27%" />
<col width="18%" />
<col width="12%" />
<col width="42%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Field</th>
<th class="head" colspan="2">Type</th>
<th class="head">Contains</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>result</td>
<td colspan="2">Dseqrecord</td>
<td>Assembeled sequence</td>
</tr>
<tr class="row-odd"><td>fragments</td>
<td colspan="2">list of Dseqrecords</td>
<td>list of fragments</td>
</tr>
<tr class="row-even"><td>offsets</td>
<td colspan="2">list of integers</td>
<td>list of cumulative offsets
aligning each source
fragment</td>
</tr>
<tr class="row-odd"><td>overlap_sizes</td>
<td colspan="2">list of integers</td>
<td>list of the length of each
overlap joining the
sequences</td>
</tr>
</tbody>
</table>
<p>Source fragments are the double stranded blunt DNA fragments 
originally added as the dsrecs argument. The parwise overlaps 
found are added ar features to the source fragments.</p>
<p>Offsets is a list with the cumulative stagger between 
each fragment.</p>
<div class="last highlight-python"><pre>            &lt;--6-&gt;        &lt;--6-&gt;        overlap_sizes = [6,6]
            
cggcggcgggcc                            \
gccgccgcccggACGGAG                      | 
                                        |
            TGCCTCaccattgc              | sticky_fragments = [12,26]
                  tggtaacgTTTTTT        |
                                        | 
                          AAAAAAcatcata | 
                                gtagtat /
                                
&lt;----------&gt;
&lt;------------------------&gt;              offsets</pre>
</div>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pydna.assembly2.recombination_assembly">
<tt class="descclassname">pydna.assembly2.</tt><tt class="descname">recombination_assembly</tt><big>(</big><em>dsrecs</em>, <em>limit=25</em><big>)</big><a class="reference internal" href="_modules/pydna/assembly2.html#recombination_assembly"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly2.recombination_assembly" title="Permalink to this definition">¶</a></dt>
<dd><p>Accepts a list of Dseqrecords and tries to assemble them into
linear and circular assemblies based on shared regions of homology 
with a minimum length given by limit.</p>
<dl class="docutils">
<dt>dsrecs <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd>a list of Dseqrecord objects.</dd>
<dt>limit <span class="classifier-delimiter">:</span> <span class="classifier">int, optional</span></dt>
<dd>limit is set to 25 nucleotides by default.</dd>
</dl>
<dl class="docutils">
<dt>linear, circular: tuple</dt>
<dd><p class="first">Linear and circular are two lists containing linear 
and circular assemblies.</p>
<p>Each object in linear or circular is a named tuple with the following
fields</p>
<table border="1" class="docutils">
<colgroup>
<col width="27%" />
<col width="18%" />
<col width="12%" />
<col width="42%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Field</th>
<th class="head" colspan="2">Type</th>
<th class="head">Contains</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>result</td>
<td colspan="2">Dseqrecord</td>
<td>Assembeled sequence</td>
</tr>
<tr class="row-odd"><td>source_fragments</td>
<td colspan="2">list of Dseqrecords</td>
<td>list of fragments</td>
</tr>
<tr class="row-even"><td>sticky_fragments</td>
<td colspan="2">list of Dseqrecords</td>
<td>list of processed fragments
with a single stranded 5&#8217;
cohesive end</td>
</tr>
<tr class="row-odd"><td>source_offsets</td>
<td colspan="2">list of integers</td>
<td>list of cumulative offsets
aligning each source
fragment</td>
</tr>
<tr class="row-even"><td>sticky_offsets</td>
<td colspan="2">list of integers</td>
<td>list of cumulative offsets
aligning each sticky
fragment</td>
</tr>
<tr class="row-odd"><td>overlap_sizes</td>
<td colspan="2">list of integers</td>
<td>list of the length of each
overlap joining the
sequences</td>
</tr>
</tbody>
</table>
<p>Source fragments are the double stranded blunt DNA fragments 
originally added as the dsrecs argument. The parwise overlaps 
found are added ar features to the source fragments.</p>
<p>Sticky fragments are source fragments that have been processed 
so that they can be ligated together by the homologous repair DNA
machinery (single-strand annealing pathway). This involves trimming 
DNA fragments so that the fragments are flanked by a 5&#8217; single stranded
overhang.</p>
<p>Source_offsets is a list with the cumulative stagger between 
each source fragment. Source offsets are different from sticky 
offsets when the overlapping sequences are located in the interior
of the fragments.</p>
<div class="last highlight-python"><pre>&lt;--------&gt;
&lt;---------------------&gt;                 source_offsets = [10,23]


cggcggcgggccTGCCTCtc                    \
gccgccgcccggACGGAGag                    | 
                                        |
          taTGCCTCaccattgcAAAAAAtt      |  source_fragments
          atACGGAGtggtaacgTTTTTTaa      |
                                        | 
                       aatAAAAAAcatcata | 
                       ttaTTTTTTgtagtat /

            &lt;--6-&gt;        &lt;--6-&gt;        overlap_sizes = [6,6]
         
cggcggcgggcc                             \
gccgccgcccggACGGAG                       | 
                                         |
            TGCCTCaccattgc               | sticky_fragments = [12,26]
                  tggtaacgTTTTTT         |
                                         | 
                          AAAAAAcatcata  | 
                                gtagtat  /
&lt;----------&gt;
&lt;------------------------&gt;               sticky_offsets</pre>
</div>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-pydna.download">
<span id="download-module"></span><h2><tt class="xref py py-mod docutils literal"><span class="pre">download</span></tt> Module<a class="headerlink" href="#module-pydna.download" title="Permalink to this headline">¶</a></h2>
<p>Provides a class for downloading sequences from genbank.</p>
<dl class="class">
<dt id="pydna.download.Genbank">
<em class="property">class </em><tt class="descclassname">pydna.download.</tt><tt class="descname">Genbank</tt><big>(</big><em>users_email</em>, <em>proxy=None</em>, <em>tool='pydna'</em><big>)</big><a class="reference internal" href="_modules/pydna/download.html#Genbank"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.download.Genbank" title="Permalink to this definition">¶</a></dt>
<dd><p>Class to facilitate download from genbank.</p>
<dl class="docutils">
<dt>users_email <span class="classifier-delimiter">:</span> <span class="classifier">string</span></dt>
<dd>Has to be a valid email address. You should always tell
Genbanks who you are, so that they can contact you.</dd>
<dt>proxy <span class="classifier-delimiter">:</span> <span class="classifier">string, optional</span></dt>
<dd>String containing a proxy url:
&#8220;proxy = &#8220;<a class="reference external" href="http://umiho.proxy.com:3128">http://umiho.proxy.com:3128</a>&#8220;</dd>
<dt>tool <span class="classifier-delimiter">:</span> <span class="classifier">string, optional</span></dt>
<dd>Default is &#8220;pydna&#8221;. This is to tell Genbank which tool you are
using.</dd>
</dl>
<p>import pydna
gb=pydna.Genbank(&#8220;<a class="reference external" href="mailto:me&#37;&#52;&#48;mail&#46;se">me<span>&#64;</span>mail<span>&#46;</span>se</a>&#8221;, proxy = &#8220;<a class="reference external" href="http://proxy.com:3128">http://proxy.com:3128</a>&#8221;)
gb.nucleotide(&#8220;L09137&#8221;) &lt;- this method does the downloading from genbank
SeqRecord(seq=Seq(&#8216;TCGCGCGTTTCGGTGATGACGGTGAAAACCTCT.....</p>
<dl class="method">
<dt id="pydna.download.Genbank.nucleotide">
<tt class="descname">nucleotide</tt><big>(</big><em>item</em><big>)</big><a class="reference internal" href="_modules/pydna/download.html#Genbank.nucleotide"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.download.Genbank.nucleotide" title="Permalink to this definition">¶</a></dt>
<dd><p>Download a genbank record using a Genbank object.</p>
<p>item is a string containing one genbank acession number [1]
for a nucleotide file:</p>
<div class="line-block">
<div class="line">A12345   = 1 letter  + 5 numerals</div>
<div class="line">AB123456 = 2 letters + 6 numerals</div>
</div>
<table class="docutils footnote" frame="void" id="id6" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[1]</td><td><a class="reference external" href="http://www.dsimb.inserm.fr/~fuchs/M2BI/AnalSeq/Annexes/Sequences/Accession_Numbers.htm">http://www.dsimb.inserm.fr/~fuchs/M2BI/AnalSeq/Annexes/Sequences/Accession_Numbers.htm</a></td></tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pydna.download.Genbank.test">
<tt class="descname">test</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/download.html#Genbank.test"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.download.Genbank.test" title="Permalink to this definition">¶</a></dt>
<dd><p>Test downloading the pUC19 plasmid sequence from genbank
returns True if successful. Can be used to test the proxy
and other network settings.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-pydna.editor">
<span id="editor-module"></span><h2><tt class="xref py py-mod docutils literal"><span class="pre">editor</span></tt> Module<a class="headerlink" href="#module-pydna.editor" title="Permalink to this headline">¶</a></h2>
<p>This module provides a class for opening a sequence using an editor
that accepts a path as a command line argument.</p>
<p>ApE - A plasmid Editor <a href="#id9"><span class="problematic" id="id10"><span id="id7"></span>[1]_</span></a> is such an editor.</p>
<div class="section" id="references">
<h3>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h3>
<table class="docutils footnote" frame="void" id="id8" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[1]</td><td><a class="reference external" href="http://biologylabs.utah.edu/jorgensen/wayned/ape/">http://biologylabs.utah.edu/jorgensen/wayned/ape/</a></td></tr>
</tbody>
</table>
</div>
<div class="section" id="examples">
<h3>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ape</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">Editor</span><span class="p">(</span><span class="s">&quot;tclsh8.6 /home/bjorn/.ApE/apeextractor/ApE.vfs/lib/app-AppMain/AppMain.tcl&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ape</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">)</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
<dl class="class">
<dt id="pydna.editor.Editor">
<em class="property">class </em><tt class="descclassname">pydna.editor.</tt><tt class="descname">Editor</tt><big>(</big><em>shell_command_for_editor</em>, <em>tmpdir=None</em><big>)</big><a class="reference internal" href="_modules/pydna/editor.html#Editor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.editor.Editor" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pydna.editor.Editor.open">
<tt class="descname">open</tt><big>(</big><em>*args</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/pydna/editor.html#Editor.open"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.editor.Editor.open" title="Permalink to this definition">¶</a></dt>
<dd><p>Open a sequence for editing in an editor.</p>
<dl class="docutils">
<dt>args <span class="classifier-delimiter">:</span> <span class="classifier">sequence or iterable of sequences</span></dt>
<dd>sequences to to be opened by the editor.</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
</div>
<div class="section" id="module-pydna.utils">
<span id="utils-module"></span><h2><tt class="xref py py-mod docutils literal"><span class="pre">utils</span></tt> Module<a class="headerlink" href="#module-pydna.utils" title="Permalink to this headline">¶</a></h2>
<p>This module provides miscellaneous functions.</p>
<dl class="function">
<dt id="pydna.utils.copy_features">
<tt class="descclassname">pydna.utils.</tt><tt class="descname">copy_features</tt><big>(</big><em>source_sr</em>, <em>target_sr</em>, <em>limit=10</em><big>)</big><a class="reference internal" href="_modules/pydna/utils.html#copy_features"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.copy_features" title="Permalink to this definition">¶</a></dt>
<dd><p>This function tries to copy all features in source_seq and copy
them to target_seq. Source_sr and target_sr are objects with
a features property, such as Dseqrecord or Biopython SeqRecord.</p>
<dl class="docutils">
<dt>source_seq <span class="classifier-delimiter">:</span> <span class="classifier">SeqRecord or Dseqrecord</span></dt>
<dd>The sequence to copy features from</dd>
<dt>target_seq <span class="classifier-delimiter">:</span> <span class="classifier">SeqRecord or Dseqrecord</span></dt>
<dd>The sequence to copy features to</dd>
</dl>
<dl class="docutils">
<dt>bool <span class="classifier-delimiter">:</span> <span class="classifier">True</span></dt>
<dd>This function acts on target_seq in place.
No data is returned.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pydna.utils.eq">
<tt class="descclassname">pydna.utils.</tt><tt class="descname">eq</tt><big>(</big><em>*args</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/pydna/utils.html#eq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.eq" title="Permalink to this definition">¶</a></dt>
<dd><p>Compares two or more DNA sequences for equality i.e. they
represent the same DNA molecule. Comparisons are case insensitive.</p>
<dl class="docutils">
<dt>args <span class="classifier-delimiter">:</span> <span class="classifier">iterable</span></dt>
<dd>iterable containing sequences
args can be strings, Biopython Seq or SeqRecord, Dseqrecord
or dsDNA objects.</dd>
<dt>circular <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd>Consider all molecules circular or linear</dd>
<dt>linear <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd>Consider all molecules circular or linear</dd>
</dl>
<dl class="docutils">
<dt>eq <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd>Returns True or False</dd>
</dl>
<p>Compares two or more DNA sequences for equality i.e. if they
represent the same DNA molecule.</p>
<p>Two linear sequences are considiered equal if either:</p>
<ul class="simple">
<li>They have the same sequence (case insensitive)</li>
<li>One sequence is the reverse complement of the other (case insensitive)</li>
</ul>
<p>Two circular sequences are considiered equal if:</p>
<ol class="arabic simple">
<li>They have the same length.</li>
</ol>
<p>AND</p>
<ol class="arabic simple" start="2">
<li>One sequence or its reverse complement can be found in the
concatenation of the other sequence with itself (they are circular
permutations).</li>
</ol>
<p>The topology for the comparison can be set using one of the keywords
linear or circular to True or False.</p>
<p>If circular or linear is not set, it will be deduced from the topology of
each sequence for sequences that have a linear or circular attribute
(like Dseq and Dseqrecord).</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna</span> <span class="kn">import</span> <span class="n">eq</span><span class="p">,</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span><span class="s">&quot;AAA&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span><span class="s">&quot;AAA&quot;</span><span class="p">,</span><span class="s">&quot;TTT&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span><span class="s">&quot;AAA&quot;</span><span class="p">,</span><span class="s">&quot;TTT&quot;</span><span class="p">,</span><span class="s">&quot;tTt&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span><span class="s">&quot;AAA&quot;</span><span class="p">,</span><span class="s">&quot;TTT&quot;</span><span class="p">,</span><span class="s">&quot;tTt&quot;</span><span class="p">,</span> <span class="n">linear</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s">&quot;Taaa&quot;</span><span class="p">,</span><span class="s">&quot;aTaa&quot;</span><span class="p">,</span> <span class="n">linear</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s">&quot;Taaa&quot;</span><span class="p">,</span><span class="s">&quot;aTaa&quot;</span><span class="p">,</span> <span class="n">circular</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;Taaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;aTaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">b</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">linear</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">linear</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s">&quot;ggatcc&quot;</span><span class="p">,</span><span class="s">&quot;GGATCC&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s">&quot;ggatcca&quot;</span><span class="p">,</span><span class="s">&quot;GGATCCa&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s">&quot;ggatcca&quot;</span><span class="p">,</span><span class="s">&quot;tGGATCC&quot;</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pydna.utils.shift_origin">
<tt class="descclassname">pydna.utils.</tt><tt class="descname">shift_origin</tt><big>(</big><em>seq</em>, <em>shift</em><big>)</big><a class="reference internal" href="_modules/pydna/utils.html#shift_origin"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.shift_origin" title="Permalink to this definition">¶</a></dt>
<dd><p>Shift the origin of seq which is assumed to be a circular
sequence.</p>
<dl class="docutils">
<dt>seq <span class="classifier-delimiter">:</span> <span class="classifier">string, Biopython Seq, Biopython SeqRecord, Dseq or Dseqrecord</span></dt>
<dd>sequence to be shifted.</dd>
</dl>
<dl class="docutils">
<dt>new_seq <span class="classifier-delimiter">:</span> <span class="classifier">string, Biopython Seq, Biopython SeqRecord, Dseq or Dseqrecord</span></dt>
<dd>sequence with a new origin.</dd>
</dl>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">shift_origin</span><span class="p">(</span><span class="s">&quot;taaa&quot;</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
<span class="go">&#39;aaat&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">shift_origin</span><span class="p">(</span><span class="s">&quot;taaa&quot;</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>
<span class="go">&#39;taaa&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">shift_origin</span><span class="p">(</span><span class="s">&quot;taaa&quot;</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
<span class="go">&#39;aata&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">shift_origin</span><span class="p">(</span><span class="s">&quot;gatc&quot;</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
<span class="go">&#39;tcga&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pydna.utils.sync">
<tt class="descclassname">pydna.utils.</tt><tt class="descname">sync</tt><big>(</big><em>seq</em>, <em>ref</em><big>)</big><a class="reference internal" href="_modules/pydna/utils.html#sync"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.sync" title="Permalink to this definition">¶</a></dt>
<dd><p>Synchronize two circular sequences.</p>
<dl class="docutils">
<dt>seq <span class="classifier-delimiter">:</span> <span class="classifier">string, Biopython Seq, Biopython SeqRecord, Dseq or Dseqrecord</span></dt>
<dd>sequence to be shifted.</dd>
<dt>ref <span class="classifier-delimiter">:</span> <span class="classifier">string, Biopython Seq, Biopython SeqRecord, Dseq or Dseqrecord</span></dt>
<dd>The reference sequence.</dd>
</dl>
<dl class="docutils">
<dt>sequence <span class="classifier-delimiter">:</span> <span class="classifier">string, Biopython Seq, Biopython SeqRecord, Dseq or Dseqrecord</span></dt>
<dd>sequence with a new origin.</dd>
</dl>
<p>This function tries to rotate the circular sequence seq
so that it has a maximum overlap with ref.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">sync</span><span class="p">(</span><span class="s">&quot;taaatc&quot;</span><span class="p">,</span><span class="s">&quot;aaataa&quot;</span><span class="p">)</span>
<span class="go">&#39;aaatct&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">sync</span><span class="p">(</span><span class="s">&quot;taaatc&quot;</span><span class="p">,</span><span class="s">&quot;aaataa&quot;</span><span class="p">)</span>
<span class="go">&#39;aaatct&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">sync</span><span class="p">(</span><span class="s">&quot;taaat&quot;</span><span class="p">,</span><span class="s">&quot;aaataa&quot;</span><span class="p">)</span>
<span class="go">&#39;aaatt&#39;</span>
</pre></div>
</div>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">pydna Package</a><ul>
<li><a class="reference internal" href="#module-pydna.dsdna"><tt class="docutils literal"><span class="pre">dsdna</span></tt> Module</a></li>
<li><a class="reference internal" href="#module-pydna.amplify"><tt class="docutils literal"><span class="pre">amplify</span></tt> Module</a></li>
<li><a class="reference internal" href="#module-pydna.primer_design"><tt class="docutils literal"><span class="pre">primer_design</span></tt> Module</a></li>
<li><a class="reference internal" href="#module-pydna.assembly"><tt class="docutils literal"><span class="pre">assembly</span></tt> Module</a></li>
<li><a class="reference internal" href="#module-pydna.assembly2"><tt class="docutils literal"><span class="pre">assembly2</span></tt> Module</a></li>
<li><a class="reference internal" href="#module-pydna.download"><tt class="docutils literal"><span class="pre">download</span></tt> Module</a></li>
<li><a class="reference internal" href="#module-pydna.editor"><tt class="docutils literal"><span class="pre">editor</span></tt> Module</a><ul>
<li><a class="reference internal" href="#references">References</a></li>
<li><a class="reference internal" href="#examples">Examples</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.utils"><tt class="docutils literal"><span class="pre">utils</span></tt> Module</a></li>
</ul>
</li>
</ul>

  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/pydna.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="index.html">Pydna 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Björn Johansson.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b1.
    </div>
  </body>
</html>